// @pierre/trees@0.0.1-beta.4/dist/features/dragAndDropUtils.js
//#region src/features/dragAndDropUtils.ts
let ItemDropCategory = /* @__PURE__ */ function(ItemDropCategory$1) {
	ItemDropCategory$1[ItemDropCategory$1["Item"] = 0] = "Item";
	ItemDropCategory$1[ItemDropCategory$1["ExpandedFolder"] = 1] = "ExpandedFolder";
	ItemDropCategory$1[ItemDropCategory$1["LastInGroup"] = 2] = "LastInGroup";
	return ItemDropCategory$1;
}({});
let PlacementType = /* @__PURE__ */ function(PlacementType$1) {
	PlacementType$1[PlacementType$1["ReorderAbove"] = 0] = "ReorderAbove";
	PlacementType$1[PlacementType$1["ReorderBelow"] = 1] = "ReorderBelow";
	PlacementType$1[PlacementType$1["MakeChild"] = 2] = "MakeChild";
	PlacementType$1[PlacementType$1["Reparent"] = 3] = "Reparent";
	return PlacementType$1;
}({});
const isOrderedDragTarget = (dragTarget) => "childIndex" in dragTarget;
const canDrop = (dataTransfer, target, tree) => {
	const draggedItems = tree.getState().dnd?.draggedItems;
	const config = tree.getConfig();
	if (draggedItems && !(config.canDrop?.(draggedItems, target) ?? true)) return false;
	if (draggedItems && draggedItems.some((draggedItem) => target.item.getId() === draggedItem.getId() || target.item.isDescendentOf(draggedItem.getId()))) return false;
	if (!draggedItems && dataTransfer && config.canDropForeignDragObject && !config.canDropForeignDragObject(dataTransfer, target)) return false;
	return true;
};
const getItemDropCategory = (item) => {
	if (item.isExpanded()) return ItemDropCategory.ExpandedFolder;
	if (item.getParent() && item.getIndexInParent() === item.getItemMeta().setSize - 1) return ItemDropCategory.LastInGroup;
	return ItemDropCategory.Item;
};
const getInsertionIndex = (children, childIndex, draggedItems) => {
	if (!draggedItems || draggedItems.length === 0) return childIndex;
	const draggedIds = /* @__PURE__ */ new Set();
	for (const draggedItem of draggedItems) draggedIds.add(draggedItem.getId());
	const endIndex = Math.min(childIndex, children.length);
	let numberOfDragItemsBeforeTarget = 0;
	for (let i = 0; i < endIndex; i++) if (draggedIds.has(children[i].getId())) numberOfDragItemsBeforeTarget++;
	return childIndex - numberOfDragItemsBeforeTarget;
};
const getTargetPlacement = (e, item, tree, canMakeChild) => {
	const config = tree.getConfig();
	if (!config.canReorder) return canMakeChild ? { type: PlacementType.MakeChild } : { type: PlacementType.ReorderBelow };
	const bb = item.getElement()?.getBoundingClientRect();
	const topPercent = bb ? (e.clientY - bb.top) / bb.height : .5;
	const leftPixels = bb ? e.clientX - bb.left : 0;
	const targetDropCategory = getItemDropCategory(item);
	const reorderAreaPercentage = !canMakeChild ? .5 : config.reorderAreaPercentage ?? .3;
	const indent = config.indent ?? 20;
	const makeChildType = canMakeChild ? PlacementType.MakeChild : PlacementType.ReorderBelow;
	if (targetDropCategory === ItemDropCategory.ExpandedFolder) {
		if (topPercent < reorderAreaPercentage) return { type: PlacementType.ReorderAbove };
		return { type: makeChildType };
	}
	if (targetDropCategory === ItemDropCategory.LastInGroup) {
		if (leftPixels < item.getItemMeta().level * indent) {
			if (topPercent < .5) return { type: PlacementType.ReorderAbove };
			const minLevel = item.getItemBelow()?.getItemMeta().level ?? 0;
			return {
				type: PlacementType.Reparent,
				reparentLevel: Math.max(minLevel, Math.floor(leftPixels / indent))
			};
		}
	}
	if (topPercent < reorderAreaPercentage) return { type: PlacementType.ReorderAbove };
	if (topPercent > 1 - reorderAreaPercentage) return { type: PlacementType.ReorderBelow };
	return { type: makeChildType };
};
const getDragCode = (item, placement) => {
	return [
		item.getId(),
		placement.type,
		placement.type === PlacementType.Reparent ? placement.reparentLevel : 0
	].join("__");
};
const getNthParent = (item, n) => {
	if (n === item.getItemMeta().level) return item;
	return getNthParent(item.getParent(), n);
};
const getReparentTarget = (item, reparentLevel, draggedItems) => {
	const itemMeta = item.getItemMeta();
	const reparentedTarget = getNthParent(item, reparentLevel - 1);
	const targetIndex = getNthParent(item, reparentLevel).getIndexInParent() + 1;
	return {
		item: reparentedTarget,
		childIndex: targetIndex,
		insertionIndex: getInsertionIndex(reparentedTarget.getChildren(), targetIndex, draggedItems),
		dragLineIndex: itemMeta.index + 1,
		dragLineLevel: reparentLevel
	};
};
const getDragTarget = (e, item, tree, canReorder = tree.getConfig().canReorder) => {
	const draggedItems = tree.getState().dnd?.draggedItems;
	const parent = item.getParent();
	const itemTarget = { item };
	const parentTarget = parent ? { item: parent } : null;
	const canBecomeSibling = parentTarget && canDrop(e.dataTransfer ?? null, parentTarget, tree);
	const placement = getTargetPlacement(e, item, tree, canDrop(e.dataTransfer ?? null, itemTarget, tree));
	if (!canReorder && parent && canBecomeSibling && placement.type !== PlacementType.MakeChild) {
		if (draggedItems?.some((draggedItem) => item.getId() === draggedItem.getId() || item.isDescendentOf(draggedItem.getId()))) return itemTarget;
		return parentTarget;
	}
	if (!canReorder && parent && !canBecomeSibling) return getDragTarget(e, parent, tree, false);
	if (!parent) return itemTarget;
	if (placement.type === PlacementType.MakeChild) return itemTarget;
	if (!canBecomeSibling) return getDragTarget(e, parent, tree, false);
	if (placement.type === PlacementType.Reparent) return getReparentTarget(item, placement.reparentLevel, draggedItems);
	const maybeAddOneForBelow = placement.type === PlacementType.ReorderAbove ? 0 : 1;
	const childIndex = item.getIndexInParent() + maybeAddOneForBelow;
	const itemMeta = item.getItemMeta();
	return {
		item: parent,
		dragLineIndex: itemMeta.index + maybeAddOneForBelow,
		dragLineLevel: itemMeta.level,
		childIndex,
		insertionIndex: getInsertionIndex(parent.getChildren(), childIndex, draggedItems)
	};
};

//#endregion
export { ItemDropCategory, PlacementType, canDrop, getDragCode, getDragTarget, getInsertionIndex, getItemDropCategory, getReparentTarget, getTargetPlacement, isOrderedDragTarget };
