// @pierre/trees@0.0.1-beta.4/dist/features/fileTreeSearchFeature.js
import { FLATTENED_PREFIX } from "../constants.js";
import { makeStateUpdater } from "@headless-tree/core";

//#region src/features/fileTreeSearchFeature.ts
const isBuiltInSearchInputEnabled = (tree) => tree.getConfig().search === true;
const defaultSearchMatcher = (search, item) => {
	if (search.length === 0) return false;
	const normalizedSearch = normalizeSearchText(search);
	if (normalizedSearch.length === 0) return false;
	const normalizedPath = getNormalizedItemPath(item);
	if (normalizedPath.includes(normalizedSearch)) return true;
	return isFuzzySubsequenceMatch(normalizedSearch, normalizedPath);
};
const normalizeSearchText = (value) => value.trim().toLowerCase().replaceAll("\\", "/");
const getNormalizedItemPath = (item) => {
	const itemPath = item.getItemData()?.path;
	return normalizeTreePath(itemPath != null && itemPath.length > 0 ? itemPath : item.getItemName());
};
const normalizeTreePath = (path) => normalizeSearchText(path.startsWith(FLATTENED_PREFIX) ? path.slice(FLATTENED_PREFIX.length) : path);
const isFuzzySubsequenceMatch = (search, target) => {
	let searchIndex = 0;
	let targetIndex = 0;
	while (searchIndex < search.length && targetIndex < target.length) {
		if (search[searchIndex] === target[targetIndex]) searchIndex += 1;
		targetIndex += 1;
	}
	return searchIndex === search.length;
};
const getSearchMode = (tree) => tree.getConfig().fileTreeSearchMode ?? "hide-non-matches";
const buildSearchIndex = (tree, rootItemId) => {
	const orderedIds = [];
	const indexById = /* @__PURE__ */ new Map();
	const parentById = /* @__PURE__ */ new Map();
	const walk = (parentId) => {
		const children = tree.retrieveChildrenIds(parentId) ?? [];
		for (const childId of children) {
			parentById.set(childId, parentId);
			indexById.set(childId, orderedIds.length);
			orderedIds.push(childId);
			walk(childId);
		}
	};
	walk(rootItemId);
	return {
		orderedIds,
		indexById,
		parentById
	};
};
const getSearchCache = (tree) => {
	const dataRef = tree.getDataRef();
	const config = tree.getConfig();
	const search = tree.getSearchValue();
	const matcher = config.isSearchMatchingItem ?? defaultSearchMatcher;
	const rootItemId = config.rootItemId;
	const dataLoader = config.dataLoader;
	const cached = dataRef.current.searchCache;
	if (cached != null && cached.search === search && cached.matcher === matcher && cached.rootItemId === rootItemId && cached.dataLoader === dataLoader) return cached;
	const index = buildSearchIndex(tree, rootItemId);
	const matchIds = search.length > 0 ? index.orderedIds.filter((itemId) => matcher(search, tree.getItemInstance(itemId))) : [];
	const matchItems = matchIds.map((itemId) => tree.getItemInstance(itemId));
	const matchIdSet = new Set(matchIds);
	const visibleIdSet = new Set(matchIds);
	for (const matchId of matchIds) addAncestorFolders(index.parentById, rootItemId, matchId, visibleIdSet);
	const nextCache = {
		search,
		rootItemId,
		dataLoader,
		matcher,
		index,
		matchItems,
		matchIds,
		matchIdSet,
		visibleIdSet
	};
	dataRef.current.searchCache = nextCache;
	dataRef.current.matchingItems = matchItems;
	return nextCache;
};
const addAncestorFolders = (parentById, rootItemId, itemId, expandedItems) => {
	let parentId = parentById.get(itemId);
	while (parentId != null && parentId !== rootItemId) {
		expandedItems.add(parentId);
		parentId = parentById.get(parentId);
	}
};
/** All folder IDs from the search index (items that have children). */
const getAllFolderIds = (tree, index) => index.orderedIds.filter((id) => (tree.retrieveChildrenIds(id) ?? []).length > 0);
const expandForMatches = (tree, cache, baselineExpandedItems, expandMatchingFolders) => {
	const expandedItems = new Set(baselineExpandedItems);
	const { parentById } = cache.index;
	for (const matchId of cache.matchIds) {
		if (expandMatchingFolders) {
			if ((tree.retrieveChildrenIds(matchId) ?? []).length > 0) expandedItems.add(matchId);
		}
		addAncestorFolders(parentById, cache.rootItemId, matchId, expandedItems);
	}
	tree.applySubStateUpdate("expandedItems", [...expandedItems]);
	tree.rebuildTree();
};
const restoreExpandedItems = (tree, previousExpandedItems, keepSelectedOpen) => {
	if (previousExpandedItems == null) return;
	const expandedItems = new Set(previousExpandedItems);
	if (keepSelectedOpen) {
		const index = buildSearchIndex(tree, tree.getConfig().rootItemId);
		for (const selectedItem of tree.getSelectedItems()) addAncestorFolders(index.parentById, tree.getConfig().rootItemId, selectedItem.getId(), expandedItems);
	}
	tree.applySubStateUpdate("expandedItems", [...expandedItems]);
	tree.rebuildTree();
};
const fileTreeSearchFeature = {
	key: "file-tree-search",
	getInitialState: (initialState) => ({
		search: null,
		...initialState
	}),
	getDefaultConfig: (defaultConfig, tree) => ({
		setSearch: makeStateUpdater("search", tree),
		isSearchMatchingItem: defaultSearchMatcher,
		...defaultConfig
	}),
	stateHandlerNames: { search: "setSearch" },
	treeInstance: {
		setSearch: ({ tree }, search) => {
			const previousSearch = tree.getState().search;
			const dataRef = tree.getDataRef();
			tree.applySubStateUpdate("search", search);
			if (previousSearch == null && search != null) dataRef.current.previousExpandedItems = tree.getState().expandedItems;
			if (search == null) {
				restoreExpandedItems(tree, dataRef.current.previousExpandedItems, true);
				dataRef.current.previousExpandedItems = null;
				dataRef.current.searchCache = void 0;
				dataRef.current.matchingItems = [];
				tree.updateDomFocus();
				return;
			}
			if (search.length === 0) {
				restoreExpandedItems(tree, dataRef.current.previousExpandedItems, false);
				dataRef.current.searchCache = void 0;
				dataRef.current.matchingItems = [];
				return;
			}
			const cache = getSearchCache(tree);
			const searchMode = getSearchMode(tree);
			const isInitialReapply = dataRef.current.previousExpandedItems == null && previousSearch === search;
			expandForMatches(tree, cache, searchMode === "expand-matches" ? isInitialReapply ? getAllFolderIds(tree, cache.index) : dataRef.current.previousExpandedItems ?? tree.getState().expandedItems : [], searchMode === "expand-matches");
			cache.matchItems[0]?.setFocused();
			if (!isInitialReapply) cache.matchItems[0]?.scrollTo({
				block: "nearest",
				inline: "nearest"
			});
		},
		openSearch: ({ tree }, initialValue = "") => {
			tree.setSearch(initialValue);
			tree.getConfig().onOpenSearch?.();
			setTimeout(() => {
				tree.getDataRef().current.searchInput?.focus();
			});
		},
		closeSearch: ({ tree }) => {
			tree.setSearch(null);
			tree.getConfig().onCloseSearch?.();
		},
		isSearchOpen: ({ tree }) => tree.getState().search !== null,
		getSearchValue: ({ tree }) => tree.getState().search ?? "",
		registerSearchInputElement: ({ tree }, element) => {
			const dataRef = tree.getDataRef();
			dataRef.current.searchInput = element;
			if (element != null && dataRef.current.keydownHandler != null) element.addEventListener("keydown", dataRef.current.keydownHandler);
		},
		getSearchInputElement: ({ tree }) => tree.getDataRef().current.searchInput ?? null,
		getSearchInputElementProps: ({ tree }) => ({
			value: tree.getSearchValue(),
			onChange: (event) => {
				const target = event.target;
				tree.setSearch(target?.value ?? "");
			},
			onBlur: () => tree.closeSearch(),
			ref: tree.registerSearchInputElement
		}),
		getSearchMatchingItems: ({ tree }) => {
			if (!tree.isSearchOpen()) return [];
			return getSearchCache(tree).matchItems;
		}
	},
	itemInstance: {
		isMatchingSearch: ({ tree, item }) => {
			if (!tree.isSearchOpen()) return false;
			return getSearchCache(tree).matchIdSet.has(item.getId());
		},
		getProps: ({ tree, prev }) => {
			const props = prev?.();
			return {
				...props,
				onMouseDown: (e) => {
					if (tree.isSearchOpen()) e.preventDefault();
					props?.onMouseDown?.(e);
				},
				onClick: (e) => {
					const shouldCloseSearch = tree.isSearchOpen();
					props?.onClick?.(e);
					if (shouldCloseSearch) tree.closeSearch();
				}
			};
		}
	},
	hotkeys: {
		openSearch: {
			hotkey: "LetterOrNumber",
			preventDefault: true,
			isEnabled: (tree) => isBuiltInSearchInputEnabled(tree) && !tree.isSearchOpen(),
			handler: (event, tree) => {
				event.stopPropagation();
				tree.openSearch(event.key);
			}
		},
		closeSearch: {
			hotkey: "Escape",
			allowWhenInputFocused: true,
			isEnabled: (tree) => tree.isSearchOpen(),
			handler: (_event, tree) => {
				tree.closeSearch();
			}
		},
		submitSearch: {
			hotkey: "Enter",
			allowWhenInputFocused: true,
			isEnabled: (tree) => tree.isSearchOpen(),
			handler: (_event, tree) => {
				tree.setSelectedItems([tree.getFocusedItem().getId()]);
				tree.closeSearch();
			}
		},
		nextSearchItem: {
			hotkey: "ArrowDown",
			allowWhenInputFocused: true,
			canRepeat: true,
			isEnabled: (tree) => tree.isSearchOpen(),
			handler: (_event, tree) => {
				const cache = getSearchCache(tree);
				const focusedId = tree.getFocusedItem().getId();
				const focusedIndex = cache.index.indexById.get(focusedId) ?? -1;
				const nextMatchId = cache.matchIds.find((matchId) => {
					return (cache.index.indexById.get(matchId) ?? -1) > focusedIndex;
				});
				if (nextMatchId) {
					const item = tree.getItemInstance(nextMatchId);
					item.setFocused();
					item.scrollTo({
						block: "nearest",
						inline: "nearest"
					});
				}
			}
		},
		previousSearchItem: {
			hotkey: "ArrowUp",
			allowWhenInputFocused: true,
			canRepeat: true,
			isEnabled: (tree) => tree.isSearchOpen(),
			handler: (_event, tree) => {
				const cache = getSearchCache(tree);
				const focusedId = tree.getFocusedItem().getId();
				const focusedIndex = cache.index.indexById.get(focusedId) ?? -1;
				for (let i = cache.matchIds.length - 1; i >= 0; i -= 1) {
					const matchId = cache.matchIds[i];
					if ((cache.index.indexById.get(matchId) ?? -1) < focusedIndex) {
						const item = tree.getItemInstance(matchId);
						item.setFocused();
						item.scrollTo({
							block: "nearest",
							inline: "nearest"
						});
						break;
					}
				}
			}
		}
	}
};
/**
* Returns the set of item IDs that should be visible when `hide-non-matches`
* search mode is active. Returns `null` when no filtering is needed.
*/
const getSearchVisibleIdSet = (tree) => {
	if (!tree.isSearchOpen()) return null;
	if (getSearchMode(tree) !== "hide-non-matches") return null;
	const cache = getSearchCache(tree);
	if (cache.matchIds.length === 0) return null;
	return cache.visibleIdSet;
};

//#endregion
export { fileTreeSearchFeature, getSearchVisibleIdSet };
