// @pierre/trees@0.0.1-beta.4/dist/features/gitStatusFeature.js
import { FLATTENED_PREFIX } from "../constants.js";
import { getGitStatusSignature } from "../utils/getGitStatusSignature.js";

//#region src/features/gitStatusFeature.ts
const normalizeGitPath = (path) => path.startsWith(FLATTENED_PREFIX) ? path.slice(FLATTENED_PREFIX.length) : path;
const getParentPath = (path) => {
	const slash = path.lastIndexOf("/");
	if (slash <= 0) return null;
	return path.slice(0, slash);
};
const addFolderIdsByPath = (pathToId, path, foldersWithChanges) => {
	const regularId = pathToId.get(path);
	if (regularId != null) foldersWithChanges.add(regularId);
	const flattenedId = pathToId.get(`${FLATTENED_PREFIX}${path}`);
	if (flattenedId != null) foldersWithChanges.add(flattenedId);
};
const buildPathToIdMapFromTree = (tree) => {
	const pathToId = /* @__PURE__ */ new Map();
	const rootItemId = tree.getConfig().rootItemId;
	const rootPath = tree.getItemInstance(rootItemId).getItemData()?.path;
	if (rootPath != null) pathToId.set(rootPath, rootItemId);
	const walk = (parentId) => {
		const children = tree.retrieveChildrenIds(parentId) ?? [];
		for (const childId of children) {
			const childPath = tree.getItemInstance(childId).getItemData()?.path;
			if (childPath != null) pathToId.set(childPath, childId);
			walk(childId);
		}
	};
	walk(rootItemId);
	return pathToId;
};
const buildGitStatusCache = (tree, gitStatus, gitStatusSignature, gitStatusPathToId) => {
	const config = tree.getConfig();
	const pathToId = gitStatusPathToId ?? buildPathToIdMapFromTree(tree);
	const foldersWithChanges = /* @__PURE__ */ new Set();
	const statusById = /* @__PURE__ */ new Map();
	for (const entry of gitStatus) {
		const normalizedPath = normalizeGitPath(entry.path);
		const idCandidates = [
			pathToId.get(entry.path),
			pathToId.get(normalizedPath),
			pathToId.get(`${FLATTENED_PREFIX}${normalizedPath}`)
		];
		for (const id of idCandidates) if (id != null) statusById.set(id, entry.status);
		let ancestorPath = getParentPath(normalizedPath);
		while (ancestorPath != null) {
			addFolderIdsByPath(pathToId, ancestorPath, foldersWithChanges);
			ancestorPath = getParentPath(ancestorPath);
		}
	}
	return {
		gitStatusSignature,
		gitStatusPathToId,
		dataLoader: config.dataLoader,
		statusById,
		foldersWithChanges
	};
};
const getGitStatusCache = (tree) => {
	const config = tree.getConfig();
	const gitStatus = config.gitStatus;
	if (gitStatus == null || gitStatus.length === 0) return null;
	const gitStatusPathToId = config.gitStatusPathToId;
	const gitStatusSignature = config.gitStatusSignature ?? getGitStatusSignature(gitStatus);
	const dataRef = tree.getDataRef();
	const cached = dataRef.current.gitStatusCache;
	if (cached != null && cached.gitStatusSignature === gitStatusSignature && cached.dataLoader === tree.getConfig().dataLoader && cached.gitStatusPathToId === gitStatusPathToId) return cached;
	const cache = buildGitStatusCache(tree, gitStatus, gitStatusSignature, gitStatusPathToId);
	dataRef.current.gitStatusCache = cache;
	return cache;
};
/**
* Returns the gitStatus statusById map for use in rendering (Root.tsx).
*/
const getGitStatusMap = (tree) => {
	return getGitStatusCache(tree);
};
const gitStatusFeature = {
	key: "git-status",
	itemInstance: {
		getGitStatus: ({ tree, item }) => {
			const cache = getGitStatusCache(tree);
			if (cache == null) return null;
			return cache.statusById.get(item.getId()) ?? null;
		},
		containsGitChange: ({ tree, item }) => {
			const cache = getGitStatusCache(tree);
			if (cache == null) return false;
			return cache.foldersWithChanges.has(item.getId());
		}
	}
};

//#endregion
export { getGitStatusMap, gitStatusFeature };
