// @pierre/trees@0.0.1-beta.4/dist/react/utils/useFileTreeInstance.js
import { getGitStatusSignature } from "../../utils/getGitStatusSignature.js";
import { FileTree } from "../../FileTree.js";
import { useCallback, useEffect, useRef } from "react";

//#region src/react/utils/useFileTreeInstance.ts
function useFileTreeInstance({ options, initialFiles, files, onFilesChange, initialExpandedItems, initialSelectedItems, initialSearchQuery, expandedItems, selectedItems, onExpandedItemsChange, onSelectedItemsChange, onSelection, gitStatus }) {
	const containerRef = useRef(null);
	const instanceRef = useRef(null);
	const syncedGitStatusSignatureRef = useRef(getGitStatusSignature(gitStatus));
	const statePropsRef = useRef({
		files,
		initialFiles,
		onFilesChange,
		expandedItems,
		selectedItems,
		onExpandedItemsChange,
		onSelectedItemsChange,
		onSelection,
		initialExpandedItems,
		initialSelectedItems,
		gitStatus,
		initialSearchQuery
	});
	statePropsRef.current = {
		files,
		initialFiles,
		onFilesChange,
		expandedItems,
		selectedItems,
		onExpandedItemsChange,
		onSelectedItemsChange,
		onSelection,
		initialExpandedItems,
		initialSelectedItems,
		gitStatus,
		initialSearchQuery
	};
	const ref = useCallback((fileTreeContainer) => {
		if (fileTreeContainer == null) {
			instanceRef.current?.cleanUp();
			instanceRef.current = null;
			containerRef.current = null;
			return;
		}
		const getExistingFileTreeId = () => {
			return Array.from(fileTreeContainer.shadowRoot?.children ?? []).find((child) => child instanceof HTMLElement && child.dataset?.fileTreeId != null && child.dataset.fileTreeId.length > 0)?.dataset?.fileTreeId;
		};
		const clearExistingFileTree = () => {
			const fileTreeElement = Array.from(fileTreeContainer.shadowRoot?.children ?? []).find((child) => child instanceof HTMLElement && child.dataset?.fileTreeId != null && child.dataset.fileTreeId.length > 0);
			if (fileTreeElement != null) fileTreeElement.replaceChildren();
		};
		const createInstance = (existingId) => {
			const sp = statePropsRef.current;
			const optionsWithFiles = options;
			syncedGitStatusSignatureRef.current = getGitStatusSignature(sp.gitStatus);
			return new FileTree({
				...options,
				initialFiles: sp.initialFiles ?? sp.files ?? optionsWithFiles.initialFiles ?? [],
				id: existingId,
				...sp.gitStatus != null && { gitStatus: sp.gitStatus }
			}, {
				initialExpandedItems: sp.initialExpandedItems ?? sp.expandedItems,
				initialSelectedItems: sp.initialSelectedItems ?? sp.selectedItems,
				initialSearchQuery: sp.initialSearchQuery,
				onExpandedItemsChange: sp.onExpandedItemsChange,
				onSelectedItemsChange: sp.onSelectedItemsChange,
				onSelection: sp.onSelection,
				onFilesChange: sp.onFilesChange
			});
		};
		const setupControlledDnD = (inst) => {
			const sp = statePropsRef.current;
			if (sp.files !== void 0 && options.dragAndDrop === true) inst.setCallbacks({ _onDragMoveFiles: (newFiles) => {
				sp.onFilesChange?.(newFiles);
			} });
		};
		const existingFileTreeId = getExistingFileTreeId();
		if (containerRef.current === fileTreeContainer && instanceRef.current != null) {
			instanceRef.current?.cleanUp();
			clearExistingFileTree();
			instanceRef.current = createInstance(existingFileTreeId);
			setupControlledDnD(instanceRef.current);
			instanceRef.current.render({ fileTreeContainer });
		} else {
			containerRef.current = fileTreeContainer;
			const hasPrerenderedContent = existingFileTreeId != null;
			instanceRef.current = createInstance(existingFileTreeId);
			setupControlledDnD(instanceRef.current);
			if (hasPrerenderedContent) instanceRef.current.hydrate({ fileTreeContainer });
			else instanceRef.current.render({ fileTreeContainer });
		}
		return () => {
			instanceRef.current?.cleanUp();
			instanceRef.current = null;
			containerRef.current = null;
		};
	}, [options]);
	useEffect(() => {
		if (files !== void 0 && instanceRef.current != null) instanceRef.current.setFiles(files);
	}, [files]);
	useEffect(() => {
		if (expandedItems !== void 0 && instanceRef.current != null) instanceRef.current.setExpandedItems(expandedItems);
	}, [expandedItems]);
	useEffect(() => {
		if (selectedItems !== void 0 && instanceRef.current != null) instanceRef.current.setSelectedItems(selectedItems);
	}, [selectedItems]);
	const gitStatusSignature = getGitStatusSignature(gitStatus);
	useEffect(() => {
		const instance = instanceRef.current;
		if (instance == null) return;
		if (syncedGitStatusSignatureRef.current === gitStatusSignature) return;
		syncedGitStatusSignatureRef.current = gitStatusSignature;
		instance.setGitStatus(gitStatus);
	}, [gitStatus, gitStatusSignature]);
	useEffect(() => {
		instanceRef.current?.setCallbacks({
			onExpandedItemsChange,
			onSelectedItemsChange,
			onSelection,
			onFilesChange,
			...files !== void 0 && options.dragAndDrop === true && { _onDragMoveFiles: (newFiles) => {
				onFilesChange?.(newFiles);
			} }
		});
	}, [
		onExpandedItemsChange,
		onSelectedItemsChange,
		onSelection,
		onFilesChange,
		files,
		options.dragAndDrop
	]);
	return { ref };
}

//#endregion
export { useFileTreeInstance };
