// @pierre/diffs@1.1.1/dist/utils/parseMergeConflictDiffFromFile.js
import { splitFileContents } from "./splitFileContents.js";
import { processFile } from "./parsePatchFiles.js";
import { getMergeConflictActionLineNumber, getMergeConflictParseResult } from "./getMergeConflictLineTypes.js";

//#region src/utils/parseMergeConflictDiffFromFile.ts
function getMergeConflictActionAnchor(action) {
	if (action.incomingLineNumber != null) return {
		side: "additions",
		lineNumber: action.incomingLineNumber
	};
	if (action.currentLineNumber != null) return {
		side: "deletions",
		lineNumber: action.currentLineNumber
	};
}
function parseMergeConflictDiffFromFile(file) {
	const lines = splitFileContents(file.contents);
	const { lineTypes, regions } = getMergeConflictParseResult(lines);
	const currentContentChunks = [];
	const incomingContentChunks = [];
	const patchContentChunks = [];
	const actions = new Array(regions.length);
	const actionOriginalLineNumbersByRegion = new Array(regions.length);
	const actionOriginalLineIndexesByRegion = new Array(regions.length);
	const actionLineIndexSet = /* @__PURE__ */ new Set();
	for (let regionIndex = 0; regionIndex < regions.length; regionIndex++) {
		const actionOriginalLineNumber = getMergeConflictActionLineNumber(regions[regionIndex]);
		const actionOriginalLineIndex = actionOriginalLineNumber - 1;
		actionOriginalLineNumbersByRegion[regionIndex] = actionOriginalLineNumber;
		actionOriginalLineIndexesByRegion[regionIndex] = actionOriginalLineIndex;
		actionLineIndexSet.add(actionOriginalLineIndex);
	}
	const actionLineNumbersByOriginalIndex = /* @__PURE__ */ new Map();
	let currentLineNumber = 0;
	let incomingLineNumber = 0;
	let actionIndex = 0;
	let nextConflict = regions[actionIndex];
	let nextActionOriginalLineNumber = nextConflict != null ? actionOriginalLineNumbersByRegion[actionIndex] : -1;
	let nextActionOriginalLineIndex = nextConflict != null ? actionOriginalLineIndexesByRegion[actionIndex] : -1;
	for (let index = 0; index < lines.length; index++) {
		const line = lines[index];
		const lineType = lineTypes[index];
		let currentLineNumberAtIndex;
		let incomingLineNumberAtIndex;
		switch (lineType) {
			case "none":
				currentContentChunks.push(line);
				incomingContentChunks.push(line);
				patchContentChunks.push(` ${line}`);
				currentLineNumber++;
				incomingLineNumber++;
				currentLineNumberAtIndex = currentLineNumber;
				incomingLineNumberAtIndex = incomingLineNumber;
				break;
			case "current":
				currentContentChunks.push(line);
				patchContentChunks.push(`-${line}`);
				currentLineNumber++;
				currentLineNumberAtIndex = currentLineNumber;
				break;
			case "incoming":
				incomingContentChunks.push(line);
				patchContentChunks.push(`+${line}`);
				incomingLineNumber++;
				incomingLineNumberAtIndex = incomingLineNumber;
				break;
			case "base":
			case "marker-start":
			case "marker-base":
			case "marker-separator":
			case "marker-end":
				currentContentChunks.push(line);
				incomingContentChunks.push(line);
				patchContentChunks.push(` ${line}`);
				currentLineNumber++;
				incomingLineNumber++;
				currentLineNumberAtIndex = currentLineNumber;
				incomingLineNumberAtIndex = incomingLineNumber;
				break;
			default: assertNever(lineType);
		}
		if (actionLineIndexSet.has(index)) actionLineNumbersByOriginalIndex.set(index, {
			currentLineNumber: currentLineNumberAtIndex,
			incomingLineNumber: incomingLineNumberAtIndex
		});
		while (nextConflict != null && nextActionOriginalLineIndex <= index) {
			const actionLineNumbers = actionLineNumbersByOriginalIndex.get(nextActionOriginalLineIndex);
			actions[actionIndex] = {
				actionOriginalLineIndex: nextActionOriginalLineIndex,
				actionOriginalLineNumber: nextActionOriginalLineNumber,
				currentLineNumber: actionLineNumbers?.currentLineNumber,
				incomingLineNumber: actionLineNumbers?.incomingLineNumber,
				conflict: nextConflict,
				conflictIndex: nextConflict.conflictIndex
			};
			actionIndex++;
			nextConflict = regions[actionIndex];
			if (nextConflict == null) break;
			nextActionOriginalLineNumber = actionOriginalLineNumbersByRegion[actionIndex];
			nextActionOriginalLineIndex = actionOriginalLineIndexesByRegion[actionIndex];
		}
	}
	const currentContents = currentContentChunks.join("");
	const incomingContents = incomingContentChunks.join("");
	const patchContents = patchContentChunks.join("");
	const currentFile = createResolvedConflictFile(file, "current", currentContents);
	const incomingFile = createResolvedConflictFile(file, "incoming", incomingContents);
	const fileDiff = processFile(createMergeConflictPatch({
		name: file.name,
		patchContents,
		currentLineCount: currentLineNumber,
		incomingLineCount: incomingLineNumber
	}), {
		oldFile: currentFile,
		newFile: incomingFile,
		cacheKey: file.cacheKey != null ? `${file.cacheKey}:merge-conflict-diff` : void 0,
		throwOnError: true
	});
	if (fileDiff == null) throw new Error("parseMergeConflictDiffFromFile: failed to build merge conflict diff metadata");
	return {
		fileDiff,
		currentFile,
		incomingFile,
		actions
	};
}
function createMergeConflictPatch({ name, patchContents, currentLineCount, incomingLineCount }) {
	return `--- ${name}\n+++ ${name}\n@@ -${currentLineCount > 0 ? 1 : 0},${currentLineCount} +${incomingLineCount > 0 ? 1 : 0},${incomingLineCount} @@\n` + patchContents;
}
function createResolvedConflictFile(file, side, contents) {
	return {
		...file,
		contents,
		cacheKey: file.cacheKey != null ? `${file.cacheKey}:merge-conflict-${side}` : void 0
	};
}
function assertNever(value) {
	throw new Error(`parseMergeConflictDiffFromFile: unknown merge conflict line type ${String(value)}`);
}

//#endregion
export { getMergeConflictActionAnchor, parseMergeConflictDiffFromFile };
