// @pierre/trees@0.0.1-beta.4/dist/utils/fileListToTree.js
import { FLATTENED_PREFIX } from "../constants.js";
import { createIdMaps } from "./createIdMaps.js";
import { createLoaderUtils } from "./createLoaderUtils.js";
import { defaultChildrenComparator, sortChildren } from "./sortChildren.js";

//#region src/utils/fileListToTree.ts
const ROOT_ID = "root";
/**
* Converts a list of file paths into a tree structure suitable for use with FileTree.
* Generates both direct children and flattened children (single-child folder chains).
*
* Time complexity: O(n * d) where n = number of files, d = average path depth
* Space complexity: O(n * d) for storing all nodes and folder relationships
*
* @param filePaths - Array of file path strings (e.g., ['src/index.ts', 'src/utils/helper.ts'])
* @param options - Optional configuration for root node
* @returns A record mapping node IDs (hashed) to FileTreeNode objects
*   with the original path stored on each node's `path` field
*/
function fileListToTree(filePaths, options = {}) {
	const { rootId = ROOT_ID, rootName = ROOT_ID, sortComparator = defaultChildrenComparator } = options;
	const tree = {};
	const folderChildren = /* @__PURE__ */ new Map();
	folderChildren.set(rootId, /* @__PURE__ */ new Set());
	for (const filePath of filePaths) {
		const parts = filePath.split("/");
		let currentPath;
		for (let i = 0; i < parts.length; i++) {
			const part = parts[i];
			const isFile = i === parts.length - 1;
			const parentPath = currentPath ?? rootId;
			currentPath = currentPath != null ? `${currentPath}/${part}` : part;
			let parentChildren = folderChildren.get(parentPath);
			if (parentChildren == null) {
				parentChildren = /* @__PURE__ */ new Set();
				folderChildren.set(parentPath, parentChildren);
			}
			parentChildren.add(currentPath);
			if (isFile) tree[currentPath] ??= {
				name: part,
				path: currentPath
			};
			else if (!folderChildren.has(currentPath)) folderChildren.set(currentPath, /* @__PURE__ */ new Set());
		}
	}
	const isFolder = (path) => folderChildren.has(path);
	const sortChildrenArray = (children) => sortChildren(children, isFolder, sortComparator);
	const getChildrenArray = (path) => {
		const children = folderChildren.get(path);
		return children != null ? [...children] : [];
	};
	const utils = createLoaderUtils(isFolder, getChildrenArray);
	const intermediateFolders = /* @__PURE__ */ new Set();
	for (const children of folderChildren.values()) for (const child of children) {
		if (!isFolder(child)) continue;
		const flattenedEndpoint = utils.getFlattenedEndpoint(child);
		if (flattenedEndpoint == null) continue;
		const flattenedFolders = utils.collectFlattenedFolders(child, flattenedEndpoint);
		for (let i = 0; i < flattenedFolders.length - 1; i++) intermediateFolders.add(flattenedFolders[i]);
		const flattenedKey = `${FLATTENED_PREFIX}${flattenedEndpoint}`;
		if (tree[flattenedKey] != null) continue;
		const flattenedName = utils.buildFlattenedName(child, flattenedEndpoint);
		const endpointChildren = folderChildren.get(flattenedEndpoint);
		const endpointDirectChildren = endpointChildren != null ? sortChildrenArray([...endpointChildren]) : [];
		const endpointFlattenedChildren = utils.buildFlattenedChildren(endpointDirectChildren);
		tree[flattenedKey] = {
			name: flattenedName,
			path: flattenedKey,
			flattens: flattenedFolders,
			children: {
				direct: endpointDirectChildren,
				...endpointFlattenedChildren != null && { flattened: endpointFlattenedChildren }
			}
		};
	}
	for (const [path, children] of folderChildren) {
		const directChildren = sortChildrenArray([...children]);
		const flattenedChildren = intermediateFolders.has(path) ? void 0 : utils.buildFlattenedChildren(directChildren);
		if (path === rootId) tree[rootId] = {
			name: rootName,
			path: rootId,
			children: {
				direct: directChildren,
				...flattenedChildren != null && { flattened: flattenedChildren }
			}
		};
		else if (tree[path] == null) {
			const lastSlashIndex = path.lastIndexOf("/");
			tree[path] = {
				name: lastSlashIndex >= 0 ? path.slice(lastSlashIndex + 1) : path,
				path,
				children: {
					direct: directChildren,
					...flattenedChildren != null && { flattened: flattenedChildren }
				}
			};
		}
	}
	const { getIdForKey } = createIdMaps(rootId);
	const mapKey = (key) => getIdForKey(key);
	const hashedTree = {};
	const keys = Object.keys(tree).sort();
	for (const key of keys) {
		const node = tree[key];
		const mappedKey = mapKey(key);
		hashedTree[mappedKey] = {
			...node,
			...node.children != null && { children: {
				direct: node.children.direct.map(mapKey),
				...node.children.flattened != null && { flattened: node.children.flattened.map(mapKey) }
			} },
			...node.flattens != null && { flattens: node.flattens.map(mapKey) }
		};
	}
	return hashedTree;
}

//#endregion
export { fileListToTree };
