// @pierre/trees@0.0.1-beta.4/dist/utils/sortChildren.js
import { FLATTENED_PREFIX } from "../constants.js";

//#region src/utils/sortChildren.ts
function stripFlattenedPrefix(path) {
	return path.startsWith(FLATTENED_PREFIX) ? path.slice(FLATTENED_PREFIX.length) : path;
}
/**
* Extracts the name (last segment) from a path.
* Handles f:: prefixed paths by stripping the prefix first.
*/
function getNameFromPath(path) {
	const actualPath = stripFlattenedPrefix(path);
	const lastSlash = actualPath.lastIndexOf("/");
	return lastSlash >= 0 ? actualPath.slice(lastSlash + 1) : actualPath;
}
function isFolderPath(path, isFolder) {
	if (path.startsWith(FLATTENED_PREFIX)) return true;
	return isFolder(path);
}
/**
* Simple alphabetical comparator
*/
const alphabeticalChildrenComparator = (a, b) => {
	const aName = getNameFromPath(a);
	const bName = getNameFromPath(b);
	return aName.localeCompare(bName);
};
/**
* Default semantic comparator for file tree children.
* Sort order:
* 1. Folders before files
* 2. Dot-prefixed (hidden) items before others within each group
* 3. Case-insensitive alphabetical within each subgroup
*/
const defaultChildrenComparator = (a, b, isFolder) => {
	const aIsFolder = isFolderPath(a, isFolder);
	if (aIsFolder !== isFolderPath(b, isFolder)) return aIsFolder ? -1 : 1;
	const aName = getNameFromPath(a);
	const bName = getNameFromPath(b);
	const aIsDot = aName.startsWith(".");
	if (aIsDot !== bName.startsWith(".")) return aIsDot ? -1 : 1;
	return aName.toLowerCase().localeCompare(bName.toLowerCase());
};
/**
* Sorts an array of child paths using the provided comparator.
*
* @param children - Array of child paths to sort
* @param isFolder - Function to check if a path is a folder
* @param comparator - Comparator function (defaults to semantic sort)
* @returns New sorted array
*/
function sortChildren(children, isFolder, comparator = defaultChildrenComparator) {
	if (comparator === false) return [...children];
	return [...children].sort((a, b) => comparator(a, b, isFolder));
}

//#endregion
export { alphabeticalChildrenComparator, defaultChildrenComparator, sortChildren };
