// @oai/remark-math/src/remark-math/math-bracket-inline.ts
/* eslint-disable @typescript-eslint/no-this-alias */
/**
 * Forked:
 * https://github.com/micromark/micromark-extension-math
 * To support `\(...\)`
 */
import { markdownLineEnding } from "micromark-util-character";
import { codes } from "micromark-util-symbol/codes";
import { types } from "micromark-util-symbol/types";

import { debug, warn } from "./helpers";

import type { Code, Effects, Event, State, Token, TokenizeContext } from "micromark-util-types";

export function mathBracketInline() {
  return {
    tokenize: tokenizeMathText,
    resolve: resolveMathText,
    previous,
  };

  function tokenizeMathText(this: TokenizeContext, effects: Effects, ok: State, nok: State) {
    const activeOpenFence: Code[] = [];
    let activeCloseFence: Code[] = [];
    let token: Token;
    const self = this;

    return start;

    function start(code: Code) {
      debug("start", code);
      if (code !== codes.backslash) {
        warn("expected `\\`", code);
      }
      if (previous.call(self, self.previous)) {
        warn("expected correct previous", self.previous);
      }
      effects.enter("mathText");
      effects.enter("mathTextSequence");
      return openingSequence(code);
    }

    function openingSequence(code: Code) {
      debug("openingSequence", code, activeOpenFence.join(","));
      if (code === codes.backslash && activeOpenFence.length === 0) {
        effects.consume(code);
        activeOpenFence.push(code);
        return openingSequence;
      }
      if (code === codes.leftSquareBracket && activeOpenFence.length === 1) {
        effects.consume(code);
        activeOpenFence.push(code);
        return openingSequence;
      }

      if (activeOpenFence.length < 2) return nok(code);
      effects.exit("mathTextSequence");
      return gap(code);
    }

    function gap(code: Code): void | State {
      debug("gap", code);
      if (code === codes.eof) {
        return nok(code);
      }

      // Closing fence?
      // Could also be data.
      if (code === codes.backslash) {
        token = effects.enter("mathTextSequence");
        // size = 0;
        activeCloseFence = [];
        return closingSequence(code);
      }

      // Tabs don’t work, and virtual spaces don’t make sense.
      if (code === codes.space) {
        effects.enter("space");
        effects.consume(code);
        effects.exit("space");
        return gap;
      }

      if (markdownLineEnding(code)) {
        effects.enter(types.lineEnding);
        effects.consume(code);
        effects.exit(types.lineEnding);
        return gap;
      }

      // Data.
      effects.enter("mathTextData");
      return data(code);
    }

    // In math.
    function data(code: Code) {
      debug("data", code);
      if (
        code === codes.eof ||
        code === codes.space ||
        code === codes.backslash ||
        markdownLineEnding(code)
      ) {
        effects.exit("mathTextData");
        return gap(code);
      }

      effects.consume(code);
      return data;
    }

    // Closing fence.
    function closingSequence(code: Code) {
      debug("closingSequence", code, activeOpenFence.join(","));
      // More.
      if (code === codes.backslash && activeCloseFence.length === 0) {
        effects.consume(code);
        activeCloseFence.push(code);
        return closingSequence;
      }
      if (code === codes.rightSquareBracket && activeCloseFence.length === 1) {
        effects.consume(code);
        activeCloseFence.push(code);
        return closingSequence;
      }

      // Done!
      if (activeCloseFence.length === activeOpenFence.length) {
        effects.exit("mathTextSequence");
        effects.exit("mathText");
        return ok(code);
      }

      // More or less accents: mark as data.
      token.type = "mathTextData";
      return data(code);
    }
  }
}

function resolveMathText(events: Event[]) {
  let tailExitIndex = events.length - 4;
  let headEnterIndex = 3;
  let index: number;
  let enter: number | undefined;

  if (events.length < 4) return events;

  // If we start and end with an EOL or a space.
  if (
    (events[headEnterIndex]![1].type === types.lineEnding ||
      events[headEnterIndex]![1].type === "space") &&
    (events[tailExitIndex]![1].type === types.lineEnding ||
      events[tailExitIndex]![1].type === "space")
  ) {
    index = headEnterIndex;

    // And we have data.
    while (++index < tailExitIndex) {
      if (events[index]![1].type === "mathTextData") {
        // Then we have padding.
        events[tailExitIndex]![1].type = "mathTextPadding";
        events[headEnterIndex]![1].type = "mathTextPadding";
        headEnterIndex += 2;
        tailExitIndex -= 2;
        break;
      }
    }
  }

  // Merge adjacent spaces and data.
  index = headEnterIndex - 1;
  tailExitIndex++;

  while (++index <= tailExitIndex) {
    if (enter === undefined) {
      if (index !== tailExitIndex && events[index]![1].type !== types.lineEnding) {
        enter = index;
      }
    } else if (index === tailExitIndex || events[index]![1].type === types.lineEnding) {
      events[enter]![1].type = "mathTextData";

      if (index !== enter + 2) {
        events[enter]![1].end = events[index - 1]![1].end;
        events.splice(enter + 2, index - enter - 2);
        tailExitIndex -= index - enter - 2;
        index = enter + 2;
      }

      enter = undefined;
    }
  }

  return events;
}

function previous(this: TokenizeContext, code: Code) {
  // If there is a previous code, there will always be a tail.
  return (
    code !== codes.backslash ||
    this.events[this.events.length - 1]?.[1].type === types.characterEscape
  );
}
