// @oai/remark-math/src/remark-math/math-bracket-block.ts
/* eslint-disable @typescript-eslint/no-this-alias */
/**
 * Forked:
 * https://github.com/micromark/micromark-extension-math
 * To support `\[...\]`
 */
import { factorySpace } from "micromark-factory-space";
import { markdownLineEnding } from "micromark-util-character";
import { codes } from "micromark-util-symbol/codes";
import { constants } from "micromark-util-symbol/constants";
import { types } from "micromark-util-symbol/types";

import { debug } from "./helpers";

import type { Code, Construct, Effects, State, TokenizeContext } from "micromark-util-types";

export const mathBracket: Construct = {
  tokenize: tokenizeMathFenced,
  concrete: true,
};

const nonLazyLine: Construct = { tokenize: tokenizeNonLazyLine, partial: true };

function tokenizeMathFenced(this: TokenizeContext, effects: Effects, ok: State, nok: State) {
  const self = this;
  const tail = this.events[this.events.length - 1];
  const initialSize =
    tail && tail[1].type === types.linePrefix ? tail[2].sliceSerialize(tail[1], true).length : 0;
  const activeOpenFence: Code[] = [];
  return start;

  function start(code: Code) {
    if (code !== codes.backslash) {
      debug("expected `\\`", code);
    }
    effects.enter("mathFlow");
    effects.enter("mathFlowFence");
    effects.enter("mathFlowFenceSequence");
    return sequenceOpen(code);
  }

  function sequenceOpen(code: Code) {
    debug("sequenceOpen", code);
    if (code === codes.backslash) {
      effects.consume(code);
      activeOpenFence.push(code);
      return sequenceOpen;
    }

    if (code === codes.leftSquareBracket && activeOpenFence[0] === codes.backslash) {
      effects.consume(code);
      activeOpenFence.push(code);
      return sequenceOpen;
    }

    effects.exit("mathFlowFenceSequence");
    return activeOpenFence.length < 2
      ? nok(code)
      : factorySpace(effects, metaOpen, types.whitespace)(code);
  }

  function metaOpen(code: Code) {
    debug("metaOpen", code);
    if (code === codes.eof || markdownLineEnding(code)) {
      return openAfter(code);
    }

    effects.enter("mathFlowFenceMeta");
    effects.enter(types.chunkString, {
      contentType: constants.contentTypeString,
    });
    return meta(code);
  }

  function meta(code: Code) {
    debug("meta", code);
    if (code === codes.eof || markdownLineEnding(code)) {
      effects.exit(types.chunkString);
      effects.exit("mathFlowFenceMeta");
      return openAfter(code);
    }

    if (code === codes.rightSquareBracket) return nok(code);
    effects.consume(code);
    return meta;
  }

  function openAfter(code: Code) {
    debug("openAfter", code);
    effects.exit("mathFlowFence");
    return self.interrupt ? ok(code) : contentStart(code);
  }

  function contentStart(code: Code) {
    debug("contentStart", code);
    if (code === codes.eof) {
      return after(code);
    }

    if (markdownLineEnding(code)) {
      return effects.attempt(
        nonLazyLine,
        effects.attempt(
          { tokenize: tokenizeClosingFence, partial: true },
          after,
          initialSize
            ? factorySpace(effects, contentStart, types.linePrefix, initialSize + 1)
            : contentStart,
        ),
        after,
      )(code);
    }

    effects.enter("mathFlowValue");
    return contentContinue(code);
  }

  function contentContinue(code: Code): void | State {
    debug("contentContinue", code);
    if (code === codes.eof || markdownLineEnding(code)) {
      effects.exit("mathFlowValue");
      return contentStart(code);
    }

    effects.consume(code);
    return contentContinue;
  }

  /** @type {State} */
  function after(code: Code) {
    debug("after", code);
    effects.exit("mathFlow");
    return ok(code);
  }

  function tokenizeClosingFence(this: TokenizeContext, effects: Effects, ok: State, nok: State) {
    const activeCloseFence: Code[] = [];

    return factorySpace(effects, closingPrefixAfter, types.linePrefix, constants.tabSize);

    function closingPrefixAfter(code: Code) {
      debug("closingPrefixAfter", code);
      effects.enter("mathFlowFence");
      effects.enter("mathFlowFenceSequence");
      return closingSequence(code);
    }

    function closingSequence(code: Code) {
      debug("closingSequence", code);
      if (code === codes.backslash && activeCloseFence.length === 0) {
        effects.consume(code);
        activeCloseFence.push(code);
        return closingSequence;
      }

      if (code === codes.rightSquareBracket && activeCloseFence[0] === codes.backslash) {
        effects.consume(code);
        activeCloseFence.push(code);
        return closingSequence;
      }

      if (activeCloseFence < activeOpenFence) return nok(code);
      effects.exit("mathFlowFenceSequence");
      return factorySpace(effects, closingSequenceEnd, types.whitespace)(code);
    }

    function closingSequenceEnd(code: Code) {
      debug("closingSequenceEnd", code);
      if (code === codes.eof || markdownLineEnding(code)) {
        effects.exit("mathFlowFence");
        return ok(code);
      }

      return nok(code);
    }
  }
}

function tokenizeNonLazyLine(this: TokenizeContext, effects: Effects, ok: State, nok: State) {
  const self = this;

  return start;

  function start(code: Code) {
    // assert(markdownLineEnding(code), 'expected eol')
    effects.enter(types.lineEnding);
    effects.consume(code);
    effects.exit(types.lineEnding);
    return lineStart;
  }

  function lineStart(code: Code) {
    return self.parser.lazy[self.now().line] ? nok(code) : ok(code);
  }
}
