// @pierre/trees@0.0.1-beta.4/dist/react/FileTree.js
'use client';


import { FILE_TREE_TAG_NAME } from "../constants.js";
import { useFileTreeInstance } from "./utils/useFileTreeInstance.js";
import { useEffect } from "react";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";

//#region src/react/FileTree.tsx
function renderFileTreeChildren() {
	return /* @__PURE__ */ jsx(Fragment, {});
}
function templateRender(children, __html) {
	if (typeof window === "undefined" && __html != null) return /* @__PURE__ */ jsxs(Fragment, { children: [/* @__PURE__ */ jsx("template", {
		shadowrootmode: "open",
		dangerouslySetInnerHTML: { __html }
	}), children] });
	return /* @__PURE__ */ jsx(Fragment, { children });
}
function FileTree({ options, className, style, prerenderedHTML, containerId, initialFiles, files, onFilesChange, initialExpandedItems, initialSelectedItems, initialSearchQuery, expandedItems, selectedItems, onExpandedItemsChange, onSelectedItemsChange, onSelection, gitStatus }) {
	const children = renderFileTreeChildren();
	const { ref } = useFileTreeInstance({
		options,
		initialFiles,
		files,
		onFilesChange,
		initialExpandedItems,
		initialSelectedItems,
		initialSearchQuery,
		expandedItems,
		selectedItems,
		onExpandedItemsChange,
		onSelectedItemsChange,
		onSelection,
		gitStatus
	});
	useEffect(() => {
		if (containerId == null) return;
		const el = document.getElementById(containerId);
		if (!(el instanceof HTMLElement)) return;
		const cleanup = ref(el);
		return () => {
			if (typeof cleanup === "function") cleanup();
			else ref(null);
		};
	}, [containerId, ref]);
	if (containerId != null) return /* @__PURE__ */ jsx(Fragment, {});
	return /* @__PURE__ */ jsx(FILE_TREE_TAG_NAME, {
		ref,
		className,
		style,
		suppressHydrationWarning: true,
		children: templateRender(children, prerenderedHTML)
	});
}

//#endregion
export { FileTree, templateRender };
