// @pierre/trees@0.0.1-beta.4/dist/loader/lazy.js
import { FLATTENED_PREFIX } from "../constants.js";
import { createIdMaps } from "../utils/createIdMaps.js";
import { createLoaderUtils } from "../utils/createLoaderUtils.js";
import { defaultChildrenComparator, sortChildren } from "../utils/sortChildren.js";

//#region src/loader/lazy.ts
/**
* Creates a lazy data loader that computes nodes on-demand.
* Best for large trees where most folders remain collapsed.
* Tradeoff: deeper navigation may trigger incremental work and caching.
*
* @param filePaths - Array of file path strings
* @param options - Configuration options
*/
function generateLazyDataLoader(filePaths, options = {}) {
	const { flattenEmptyDirectories = false, rootId = "root", rootName = "root", sortComparator = defaultChildrenComparator } = options;
	const sortedPaths = [...filePaths].sort();
	const lowerBound = (target) => {
		let lo = 0;
		let hi = sortedPaths.length;
		while (lo < hi) {
			const mid = lo + hi >>> 1;
			if (sortedPaths[mid] < target) lo = mid + 1;
			else hi = mid;
		}
		return lo;
	};
	const folderSet = /* @__PURE__ */ new Set();
	for (const path of sortedPaths) {
		let current = "";
		const parts = path.split("/");
		for (let i = 0; i < parts.length - 1; i++) {
			current = current !== "" ? `${current}/${parts[i]}` : parts[i];
			folderSet.add(current);
		}
	}
	const nodeCache = /* @__PURE__ */ new Map();
	const directChildrenCache = /* @__PURE__ */ new Map();
	const isFolder = (path) => {
		return path === rootId || folderSet.has(path);
	};
	const getNameFromPath = (path) => {
		const lastSlash = path.lastIndexOf("/");
		return lastSlash >= 0 ? path.slice(lastSlash + 1) : path;
	};
	const findDirectChildren = (parentPath) => {
		const cached = directChildrenCache.get(parentPath);
		if (cached != null) return cached;
		const children = /* @__PURE__ */ new Set();
		if (parentPath === rootId) for (const path of sortedPaths) {
			const slashIndex = path.indexOf("/");
			const childSegment = slashIndex >= 0 ? path.slice(0, slashIndex) : path;
			if (childSegment !== "") children.add(childSegment);
		}
		else {
			const prefix = `${parentPath}/`;
			const prefixLen = prefix.length;
			for (let i = lowerBound(prefix); i < sortedPaths.length; i += 1) {
				const path = sortedPaths[i];
				if (!path.startsWith(prefix)) break;
				const relativePath = path.slice(prefixLen);
				if (relativePath === "") continue;
				const slashIndex = relativePath.indexOf("/");
				const childSegment = slashIndex >= 0 ? relativePath.slice(0, slashIndex) : relativePath;
				if (childSegment !== "") children.add(`${parentPath}/${childSegment}`);
			}
		}
		const result = sortChildren([...children], isFolder, sortComparator);
		directChildrenCache.set(parentPath, result);
		return result;
	};
	const utils = createLoaderUtils(isFolder, findDirectChildren);
	const intermediateFolders = /* @__PURE__ */ new Set();
	const flattenedKeys = /* @__PURE__ */ new Set();
	for (const folder of folderSet) {
		const flattenedEndpoint = utils.getFlattenedEndpoint(folder);
		if (flattenedEndpoint == null) continue;
		const flattenedFolders = utils.collectFlattenedFolders(folder, flattenedEndpoint);
		for (let i = 0; i < flattenedFolders.length - 1; i++) intermediateFolders.add(flattenedFolders[i]);
		flattenedKeys.add(`${FLATTENED_PREFIX}${flattenedEndpoint}`);
	}
	const { getIdForKey, getKeyForId } = createIdMaps(rootId);
	const allKeys = new Set([rootId]);
	for (const path of sortedPaths) allKeys.add(path);
	for (const folder of folderSet) allKeys.add(folder);
	for (const flattenedKey of flattenedKeys) allKeys.add(flattenedKey);
	for (const key of Array.from(allKeys).sort()) getIdForKey(key);
	const mapKey = (key) => getIdForKey(key);
	const findFlattenedChainStart = (endPath) => {
		const parts = endPath.split("/");
		for (let i = 1; i < parts.length; i++) {
			const ancestorPath = parts.slice(0, i).join("/");
			if (utils.getFlattenedEndpoint(ancestorPath) === endPath) return ancestorPath;
		}
		return endPath;
	};
	const getItem = (id) => {
		const cached = nodeCache.get(id);
		if (cached != null) return cached;
		const key = getKeyForId(id);
		if (key == null) throw new Error(`generateLazyDataLoader: unknown id ${id}`);
		let node;
		if (key === rootId) {
			const directChildren = findDirectChildren(rootId);
			const flattenedChildren = utils.buildFlattenedChildren(directChildren);
			node = {
				name: rootName,
				path: rootId,
				children: {
					direct: directChildren.map(mapKey),
					...flattenedChildren != null && { flattened: flattenedChildren.map(mapKey) }
				}
			};
		} else if (key.startsWith(FLATTENED_PREFIX)) {
			const endPath = key.slice(FLATTENED_PREFIX.length);
			const startPath = findFlattenedChainStart(endPath);
			const flattenedFolders = utils.collectFlattenedFolders(startPath, endPath);
			const directChildren = findDirectChildren(endPath);
			const flattenedChildren = utils.buildFlattenedChildren(directChildren);
			node = {
				name: utils.buildFlattenedName(startPath, endPath),
				path: key,
				flattens: flattenedFolders.map(mapKey),
				children: {
					direct: directChildren.map(mapKey),
					...flattenedChildren != null && { flattened: flattenedChildren.map(mapKey) }
				}
			};
		} else if (isFolder(key)) {
			const directChildren = findDirectChildren(key);
			const flattenedChildren = intermediateFolders.has(key) ? void 0 : utils.buildFlattenedChildren(directChildren);
			node = {
				name: getNameFromPath(key),
				path: key,
				children: {
					direct: directChildren.map(mapKey),
					...flattenedChildren != null && { flattened: flattenedChildren.map(mapKey) }
				}
			};
		} else node = {
			name: getNameFromPath(key),
			path: key
		};
		nodeCache.set(id, node);
		return node;
	};
	const getChildren = (id) => {
		const item = getItem(id);
		if (item.children == null) return [];
		if (flattenEmptyDirectories && item.children.flattened != null) return item.children.flattened;
		return item.children.direct;
	};
	return {
		getItem,
		getChildren
	};
}

//#endregion
export { generateLazyDataLoader };
