// @pierre/trees@0.0.1-beta.4/dist/FileTree.js
import { FILE_TREE_TAG_NAME, FILE_TREE_UNSAFE_CSS_ATTRIBUTE, FLATTENED_PREFIX } from "./constants.js";
import { FileTreeContainerLoaded } from "./components/web-components.js";
import { SVGSpriteSheet } from "./sprite.js";
import { wrapUnsafeCSS } from "./utils/cssWrappers.js";
import { expandImplicitParentDirectories } from "./utils/expandImplicitParentDirectories.js";
import { buildDirectChildCountMap, expandPathsWithAncestors, filterOrphanedPaths, isOrphanedPathForExpandedSet } from "./utils/expandPaths.js";
import { preactHydrateRoot, preactRenderRoot, preactUnmountRoot } from "./utils/preactRenderer.js";
import "@headless-tree/core";

//#region src/FileTree.ts
let instanceId = -1;
const isBrowser = typeof document !== "undefined";
var FileTree = class {
	static LoadedCustomComponent = FileTreeContainerLoaded;
	__id;
	fileTreeContainer;
	divWrapper;
	defaultSpriteSheet;
	unsafeCSSStyle;
	/** Populated by the Preact Root component with the tree instance + maps. */
	handleRef = { current: null };
	/** Populated by FileTree, read by the Preact Root for callbacks. */
	callbacksRef;
	expandPathsCache = /* @__PURE__ */ new Map();
	expandPathsCacheFor = null;
	childCountCache = null;
	childCountCacheFor = null;
	constructor(options, stateConfig = {}) {
		this.options = options;
		this.stateConfig = stateConfig;
		if (typeof document !== "undefined") this.fileTreeContainer = document.createElement(FILE_TREE_TAG_NAME);
		this.__id = options.id ?? `ft_${isBrowser ? "brw" : "srv"}_${++instanceId}`;
		this.callbacksRef = { current: {
			onExpandedItemsChange: stateConfig.onExpandedItemsChange,
			onSelectedItemsChange: stateConfig.onSelectedItemsChange,
			onSelection: stateConfig.onSelection,
			onFilesChange: stateConfig.onFilesChange,
			_onDragMoveFiles: options.dragAndDrop === true ? (newFiles) => this.setFiles(newFiles) : void 0
		} };
	}
	setExpandedItems(items) {
		const handle = this.handleRef.current;
		if (handle == null) return;
		if (this.expandPathsCacheFor !== handle.pathToId) {
			this.expandPathsCache.clear();
			this.expandPathsCacheFor = handle.pathToId;
		}
		if (this.childCountCacheFor !== handle.pathToId) {
			this.childCountCache = buildDirectChildCountMap(handle.pathToId);
			this.childCountCacheFor = handle.pathToId;
		}
		const desiredExpandedSet = new Set(expandImplicitParentDirectories(items));
		const currentIds = handle.tree.getState().expandedItems ?? [];
		const currentPaths = [];
		{
			const seen = /* @__PURE__ */ new Set();
			for (const id of currentIds) {
				const raw = handle.idToPath.get(id);
				if (raw == null) continue;
				const path = raw.startsWith(FLATTENED_PREFIX) ? raw.slice(FLATTENED_PREFIX.length) : raw;
				if (path === "root" || path === "") continue;
				if (seen.has(path)) continue;
				seen.add(path);
				currentPaths.push(path);
			}
		}
		const hiddenPathsToPreserve = [];
		for (const path of currentPaths) {
			if (desiredExpandedSet.has(path)) continue;
			if (isOrphanedPathForExpandedSet(path, desiredExpandedSet, handle.pathToId, {
				flattenEmptyDirectories: this.options.flattenEmptyDirectories,
				childCount: this.childCountCache ?? void 0
			})) hiddenPathsToPreserve.push(path);
		}
		const ids = expandPathsWithAncestors(items, handle.pathToId, {
			flattenEmptyDirectories: this.options.flattenEmptyDirectories,
			cache: this.expandPathsCache
		});
		const flattenEmptyDirectories = this.options.flattenEmptyDirectories === true;
		const preserveIds = hiddenPathsToPreserve.map((path) => {
			if (path.startsWith(FLATTENED_PREFIX)) return handle.pathToId.get(path);
			return flattenEmptyDirectories ? handle.pathToId.get(FLATTENED_PREFIX + path) ?? handle.pathToId.get(path) : handle.pathToId.get(path);
		}).filter((id) => id != null);
		if (preserveIds.length === 0) handle.tree.applySubStateUpdate("expandedItems", () => ids);
		else {
			const next = new Set(ids);
			for (const id of preserveIds) next.add(id);
			handle.tree.applySubStateUpdate("expandedItems", () => Array.from(next));
		}
		handle.tree.scheduleRebuildTree();
	}
	setSelectedItems(items) {
		const handle = this.handleRef.current;
		if (handle == null) return;
		const flattenEmptyDirectories = this.options.flattenEmptyDirectories === true;
		const ids = items.map((path) => {
			if (path.startsWith(FLATTENED_PREFIX)) return handle.pathToId.get(path);
			return flattenEmptyDirectories ? handle.pathToId.get(FLATTENED_PREFIX + path) ?? handle.pathToId.get(path) : handle.pathToId.get(path);
		}).filter((id) => id != null);
		handle.tree.applySubStateUpdate("selectedItems", () => ids);
	}
	expandItem(path) {
		const current = this.getExpandedItems();
		if (!current.includes(path)) this.setExpandedItems([...current, path]);
	}
	collapseItem(path) {
		const handle = this.handleRef.current;
		if (handle == null) return;
		const idsToRemove = /* @__PURE__ */ new Set();
		const id = handle.pathToId.get(path);
		if (id != null) idsToRemove.add(id);
		const flatId = handle.pathToId.get(FLATTENED_PREFIX + path);
		if (flatId != null) idsToRemove.add(flatId);
		if (idsToRemove.size === 0) return;
		const currentIds = handle.tree.getState().expandedItems ?? [];
		handle.tree.applySubStateUpdate("expandedItems", () => currentIds.filter((i) => !idsToRemove.has(i)));
		handle.tree.scheduleRebuildTree();
	}
	toggleItemExpanded(path) {
		const handle = this.handleRef.current;
		if (handle == null) return;
		const id = handle.pathToId.get(path) ?? handle.pathToId.get(FLATTENED_PREFIX + path);
		if (id == null) return;
		if ((handle.tree.getState().expandedItems ?? []).includes(id)) this.collapseItem(path);
		else this.expandItem(path);
	}
	getExpandedItems() {
		const handle = this.handleRef.current;
		if (handle == null) return [];
		return filterOrphanedPaths((handle.tree.getState().expandedItems ?? []).map((id) => handle.idToPath.get(id)).filter((path) => path != null).map((path) => path.startsWith(FLATTENED_PREFIX) ? path.slice(FLATTENED_PREFIX.length) : path), handle.pathToId, this.options.flattenEmptyDirectories);
	}
	getSelectedItems() {
		const handle = this.handleRef.current;
		if (handle == null) return [];
		return (handle.tree.getState().selectedItems ?? []).map((id) => handle.idToPath.get(id)).filter((path) => path != null).map((path) => path.startsWith(FLATTENED_PREFIX) ? path.slice(FLATTENED_PREFIX.length) : path);
	}
	setCallbacks(callbacks) {
		Object.assign(this.callbacksRef.current, callbacks);
	}
	setGitStatus(entries) {
		this.options = {
			...this.options,
			gitStatus: entries
		};
		this.rerender();
	}
	getGitStatus() {
		return this.options.gitStatus;
	}
	setFiles(files) {
		if (this.options.initialFiles === files) return;
		this.options = {
			...this.options,
			initialFiles: files
		};
		this.callbacksRef.current.onFilesChange?.(files);
		this.rerender();
	}
	getFiles() {
		return this.options.initialFiles;
	}
	setOptions(options, state) {
		if (options.dragAndDrop === false) this.callbacksRef.current._onDragMoveFiles = void 0;
		else if (options.dragAndDrop === true && this.callbacksRef.current._onDragMoveFiles == null) this.callbacksRef.current._onDragMoveFiles = (newFiles) => this.setFiles(newFiles);
		if (state?.onExpandedItemsChange !== void 0) this.callbacksRef.current.onExpandedItemsChange = state.onExpandedItemsChange;
		if (state?.onSelectedItemsChange !== void 0) this.callbacksRef.current.onSelectedItemsChange = state.onSelectedItemsChange;
		if (state?.onSelection !== void 0) this.callbacksRef.current.onSelection = state.onSelection;
		if (state?.onFilesChange !== void 0) this.callbacksRef.current.onFilesChange = state.onFilesChange;
		const structuralKeys = [
			"dragAndDrop",
			"fileTreeSearchMode",
			"gitStatus",
			"initialFiles",
			"icons",
			"flattenEmptyDirectories",
			"lockedPaths",
			"onCollision",
			"sort",
			"unsafeCSS",
			"useLazyDataLoader",
			"virtualize"
		];
		let needsRerender = false;
		for (const key of structuralKeys) if (key in options) {
			needsRerender = true;
			break;
		}
		const nextFiles = state?.files;
		const stateFilesChanged = nextFiles !== void 0 && this.options.initialFiles !== nextFiles;
		this.options = {
			...this.options,
			...options,
			...nextFiles !== void 0 && { initialFiles: nextFiles }
		};
		if (state != null) this.stateConfig = {
			...this.stateConfig,
			...state
		};
		if (needsRerender || stateFilesChanged) {
			if (stateFilesChanged && nextFiles !== void 0) this.callbacksRef.current.onFilesChange?.(nextFiles);
			this.rerender();
		} else {
			if (state?.expandedItems !== void 0) this.setExpandedItems(state.expandedItems);
			if (state?.selectedItems !== void 0) this.setSelectedItems(state.selectedItems);
		}
	}
	buildRootProps() {
		return {
			fileTreeOptions: this.options,
			stateConfig: this.stateConfig,
			handleRef: this.handleRef,
			callbacksRef: this.callbacksRef
		};
	}
	isVirtualized() {
		return this.options.virtualize != null && this.options.virtualize !== false;
	}
	syncVirtualizedLayoutAttrs(fileTreeContainer, divWrapper) {
		const host = fileTreeContainer ?? this.fileTreeContainer;
		const wrapper = divWrapper ?? this.divWrapper;
		const isVirtualized = this.isVirtualized();
		if (host != null) if (isVirtualized) host.dataset.fileTreeVirtualized = "true";
		else delete host.dataset.fileTreeVirtualized;
		if (wrapper != null) if (isVirtualized) wrapper.dataset.fileTreeVirtualizedWrapper = "true";
		else delete wrapper.dataset.fileTreeVirtualizedWrapper;
	}
	rerender() {
		if (this.divWrapper == null) return;
		if (this.fileTreeContainer != null) {
			this.syncIconSpriteSheets(this.fileTreeContainer);
			this.syncUnsafeCSS(this.fileTreeContainer);
		}
		this.syncVirtualizedLayoutAttrs(this.fileTreeContainer, this.divWrapper);
		preactRenderRoot(this.divWrapper, this.buildRootProps());
	}
	parseSpriteSheet(spriteSheet) {
		const fragment = document.createElement("div");
		fragment.innerHTML = spriteSheet;
		const svg = fragment.querySelector("svg");
		if (svg instanceof SVGElement) return svg;
	}
	isDefaultSpriteSheet(spriteSheet) {
		return spriteSheet.querySelector("#file-tree-icon-chevron") instanceof SVGElement && spriteSheet.querySelector("#file-tree-icon-file") instanceof SVGElement && spriteSheet.querySelector("#file-tree-icon-dot") instanceof SVGElement && spriteSheet.querySelector("#file-tree-icon-lock") instanceof SVGElement;
	}
	getTopLevelSpriteSheets(shadowRoot) {
		return Array.from(shadowRoot.children).filter((element) => element instanceof SVGElement);
	}
	ensureDefaultSpriteSheet(shadowRoot) {
		let defaultSprite = this.defaultSpriteSheet != null && this.defaultSpriteSheet.parentNode === shadowRoot ? this.defaultSpriteSheet : void 0;
		defaultSprite ??= this.getTopLevelSpriteSheets(shadowRoot).find((sprite) => this.isDefaultSpriteSheet(sprite));
		if (defaultSprite == null) {
			const builtInSprite = this.parseSpriteSheet(SVGSpriteSheet);
			if (builtInSprite != null) {
				shadowRoot.appendChild(builtInSprite);
				defaultSprite = builtInSprite;
			}
		}
		this.defaultSpriteSheet = defaultSprite;
	}
	syncCustomSpriteSheet(shadowRoot) {
		const topLevelSprites = this.getTopLevelSpriteSheets(shadowRoot);
		const defaultSprite = topLevelSprites.find((sprite) => this.isDefaultSpriteSheet(sprite));
		const currentCustomSprites = topLevelSprites.filter((sprite) => sprite !== defaultSprite);
		const customSpriteSheet = this.options.icons?.spriteSheet?.trim() ?? "";
		if (customSpriteSheet.length === 0) {
			for (const customSprite$1 of currentCustomSprites) customSprite$1.remove();
			return;
		}
		const customSprite = this.parseSpriteSheet(customSpriteSheet);
		if (customSprite == null) {
			for (const customSprite$1 of currentCustomSprites) customSprite$1.remove();
			return;
		}
		if (currentCustomSprites.length === 1 && currentCustomSprites[0].outerHTML === customSprite.outerHTML) return;
		for (const currentCustomSprite of currentCustomSprites) currentCustomSprite.remove();
		shadowRoot.appendChild(customSprite);
	}
	syncIconSpriteSheets(fileTreeContainer) {
		const shadowRoot = fileTreeContainer.shadowRoot;
		if (shadowRoot == null) return;
		this.ensureDefaultSpriteSheet(shadowRoot);
		this.syncCustomSpriteSheet(shadowRoot);
	}
	syncUnsafeCSS(fileTreeContainer) {
		const shadowRoot = fileTreeContainer.shadowRoot;
		if (shadowRoot == null) return;
		let unsafeStyle = this.unsafeCSSStyle instanceof HTMLStyleElement && this.unsafeCSSStyle.parentNode === shadowRoot ? this.unsafeCSSStyle : void 0;
		unsafeStyle ??= Array.from(shadowRoot.children).find((element) => element instanceof HTMLStyleElement && element.hasAttribute(FILE_TREE_UNSAFE_CSS_ATTRIBUTE));
		const unsafeCSS = this.options.unsafeCSS?.trim() ?? "";
		if (unsafeCSS.length === 0) {
			unsafeStyle?.remove();
			this.unsafeCSSStyle = void 0;
			return;
		}
		if (unsafeStyle == null) {
			unsafeStyle = document.createElement("style");
			unsafeStyle.setAttribute(FILE_TREE_UNSAFE_CSS_ATTRIBUTE, "");
			shadowRoot.appendChild(unsafeStyle);
		}
		const wrappedUnsafeCSS = wrapUnsafeCSS(unsafeCSS);
		if (unsafeStyle.textContent !== wrappedUnsafeCSS) unsafeStyle.textContent = wrappedUnsafeCSS;
		this.unsafeCSSStyle = unsafeStyle;
	}
	getOrCreateFileTreeContainer(fileTreeContainer, parentNode) {
		this.fileTreeContainer = fileTreeContainer ?? this.fileTreeContainer ?? document.createElement(FILE_TREE_TAG_NAME);
		if (parentNode != null && this.fileTreeContainer.parentNode !== parentNode) parentNode.appendChild(this.fileTreeContainer);
		if (this.fileTreeContainer.shadowRoot == null) try {
			this.fileTreeContainer.attachShadow({ mode: "open" });
		} catch {}
		this.syncIconSpriteSheets(this.fileTreeContainer);
		this.syncUnsafeCSS(this.fileTreeContainer);
		return this.fileTreeContainer;
	}
	getFileTreeContainer() {
		return this.fileTreeContainer;
	}
	getOrCreateDivWrapperNode(container) {
		if (this.divWrapper == null) {
			for (const element of Array.from(container.shadowRoot?.children ?? [])) if (element instanceof HTMLDivElement && element.dataset.fileTreeId === this.__id) {
				this.divWrapper = element;
				break;
			}
			if (this.divWrapper == null) {
				this.divWrapper = document.createElement("div");
				this.divWrapper.dataset.fileTreeId = this.__id.toString();
				container.shadowRoot?.appendChild(this.divWrapper);
			}
		} else {
			const targetParent = container.shadowRoot ?? container;
			if (this.divWrapper.parentNode !== targetParent) targetParent.appendChild(this.divWrapper);
		}
		return this.divWrapper;
	}
	render({ fileTreeContainer, containerWrapper }) {
		fileTreeContainer = this.getOrCreateFileTreeContainer(fileTreeContainer, containerWrapper);
		const divWrapper = this.getOrCreateDivWrapperNode(fileTreeContainer);
		this.syncVirtualizedLayoutAttrs(fileTreeContainer, divWrapper);
		preactRenderRoot(divWrapper, this.buildRootProps());
	}
	hydrate(props) {
		const { fileTreeContainer } = props;
		let discoveredId;
		for (const element of Array.from(fileTreeContainer.shadowRoot?.children ?? [])) {
			if (element instanceof SVGElement) continue;
			if (!(element instanceof HTMLElement)) continue;
			if (element instanceof HTMLDivElement && element.dataset.fileTreeId != null) {
				discoveredId ??= element.dataset.fileTreeId;
				if (element.dataset.fileTreeId === this.__id) {
					this.divWrapper = element;
					break;
				}
				this.divWrapper ??= element;
				continue;
			}
		}
		if (discoveredId != null && this.__id !== discoveredId) {
			this.__id = discoveredId;
			this.options = {
				...this.options,
				id: discoveredId
			};
		}
		this.fileTreeContainer = fileTreeContainer;
		this.syncIconSpriteSheets(fileTreeContainer);
		this.syncUnsafeCSS(fileTreeContainer);
		this.syncVirtualizedLayoutAttrs(fileTreeContainer, this.divWrapper);
		if (this.divWrapper == null) {
			console.warn("FileTree: expected html not found, rendering instead");
			this.render(props);
		} else {
			preactHydrateRoot(this.divWrapper, this.buildRootProps());
			if (this.options.dragAndDrop === true) for (const btn of this.divWrapper.querySelectorAll("button[data-type=\"item\"]")) btn.draggable = true;
		}
	}
	cleanUp() {
		if (this.fileTreeContainer != null) delete this.fileTreeContainer.dataset.fileTreeVirtualized;
		if (this.divWrapper != null) {
			delete this.divWrapper.dataset.fileTreeVirtualizedWrapper;
			preactUnmountRoot(this.divWrapper);
		}
		this.handleRef.current = null;
		this.expandPathsCache.clear();
		this.expandPathsCacheFor = null;
		this.unsafeCSSStyle?.remove();
		this.unsafeCSSStyle = void 0;
		this.fileTreeContainer = void 0;
		this.divWrapper = void 0;
		this.defaultSpriteSheet = void 0;
	}
};

//#endregion
export { FileTree };
