// @pierre/trees@0.0.1-beta.4/dist/components/Root.js
import { FLATTENED_PREFIX } from "../constants.js";
import { expandPathsWithAncestors, filterOrphanedPaths } from "../utils/expandPaths.js";
import { dragAndDropFeature } from "../features/dragAndDropFeature.js";
import { fileTreeSearchFeature, getSearchVisibleIdSet } from "../features/fileTreeSearchFeature.js";
import { getGitStatusSignature } from "../utils/getGitStatusSignature.js";
import { getGitStatusMap, gitStatusFeature } from "../features/gitStatusFeature.js";
import { generateLazyDataLoader } from "../loader/lazy.js";
import { fileListToTree } from "../utils/fileListToTree.js";
import { generateSyncDataLoader } from "../loader/sync.js";
import { computeNewFilesAfterDrop } from "../utils/computeNewFilesAfterDrop.js";
import { controlledExpandedPathsToExpandedIds } from "../utils/controlledExpandedState.js";
import { buildAncestorChains, buildChildToParent } from "../utils/guideLineAncestors.js";
import { useTree } from "./hooks/useTree.js";
import { Icon } from "./Icon.js";
import { VirtualizedList } from "./VirtualizedList.js";
import { expandAllFeature, hotkeysCoreFeature, keyboardDragAndDropFeature, propMemoizationFeature, selectionFeature, syncDataLoaderFeature } from "@headless-tree/core";
import { Component, Fragment, createElement } from "preact";
import { useCallback, useEffect, useLayoutEffect, useMemo, useRef } from "preact/hooks";
import { jsx, jsxs } from "preact/jsx-runtime";

//#region src/components/Root.tsx
function memo(c, comparer) {
	class Memoed extends Component {
		shouldComponentUpdate(nextProps) {
			return !comparer(this.props, nextProps);
		}
		render() {
			return createElement(c, this.props);
		}
	}
	Memoed.displayName = `Memo(${c.displayName ?? c.name ?? "Component"})`;
	return Memoed;
}
const getSelectionPath = (path) => path.startsWith(FLATTENED_PREFIX) ? path.slice(FLATTENED_PREFIX.length) : path;
const getFilesSignature = (files) => `${files.length}\0${files.join("\0")}`;
const EMPTY_ANCESTORS = [];
function FlattenedDirectoryName({ tree, idToPath, flattens, fallbackName }) {
	"use no memo";
	const segments = useMemo(() => {
		const result = [];
		for (const id of flattens) {
			const item = tree.getItemInstance(id);
			if (item != null) result.push({
				id: item.getId(),
				label: item.getItemName()
			});
			else {
				const path = idToPath.get(id);
				const label = path != null ? path.split("/").pop() ?? id : id;
				result.push({
					id,
					label
				});
			}
		}
		return result;
	}, [
		flattens,
		tree,
		idToPath
	]);
	if (segments.length === 0) return /* @__PURE__ */ jsx("span", {
		"data-item-flattened-subitems": true,
		children: fallbackName.replace(/\//g, " / ")
	});
	return /* @__PURE__ */ jsx("span", {
		"data-item-flattened-subitems": true,
		children: segments.map(({ id, label }, index) => {
			const isLast = index === segments.length - 1;
			return /* @__PURE__ */ jsxs(Fragment, { children: [/* @__PURE__ */ jsx("span", {
				"data-item-flattened-subitem": id,
				children: label
			}), !isLast ? " / " : ""] }, id);
		})
	});
}
function treeItemPropsAreEqual(prev, next) {
	return prev.itemId === next.itemId && prev.hasChildren === next.hasChildren && prev.isExpanded === next.isExpanded && prev.itemName === next.itemName && prev.level === next.level && prev.isSelected === next.isSelected && prev.isFocused === next.isFocused && prev.isSearchMatch === next.isSearchMatch && prev.isDragTarget === next.isDragTarget && prev.isDragging === next.isDragging && prev.isDnD === next.isDnD && prev.isFlattenedDirectory === next.isFlattenedDirectory && prev.isLocked === next.isLocked && prev.gitStatus === next.gitStatus && prev.containsGitChange === next.containsGitChange && prev.flattens === next.flattens && prev.ancestors === next.ancestors && prev.treeDomId === next.treeDomId;
}
function TreeItemInner({ item, tree, itemId, hasChildren, itemName, level, isSelected, isFocused, isSearchMatch, isDragTarget, isDragging, isDnD, isFlattenedDirectory, isLocked, gitStatus: itemGitStatus, containsGitChange: itemContainsGitChange, flattens, idToPath, ancestors, treeDomId, remapIcon, detectFlattenedSubfolder, clearFlattenedSubfolder }) {
	"use no memo";
	const alignCapitals = itemName.charAt(0).toUpperCase() === itemName.charAt(0);
	const selectionProps = isSelected ? { "data-item-selected": true } : {};
	const focusedProps = isFocused ? { "data-item-focused": true } : {};
	const searchMatchProps = isSearchMatch ? { "data-item-search-match": true } : {};
	const dragProps = isDnD ? {
		...isDragTarget && { "data-item-drag-target": true },
		...isDragging && { "data-item-dragging": true }
	} : {};
	const gitStatusProps = {
		...itemGitStatus != null && { "data-item-git-status": itemGitStatus },
		...itemContainsGitChange && { "data-item-contains-git-change": "true" }
	};
	const baseProps = item.getProps();
	const itemProps = isDnD && isFlattenedDirectory ? {
		...baseProps,
		onDragOver: (e) => {
			baseProps.onDragOver?.(e);
			detectFlattenedSubfolder(e);
		},
		onDragLeave: (e) => {
			clearFlattenedSubfolder();
			baseProps.onDragLeave?.(e);
		},
		onDrop: (e) => {
			baseProps.onDrop?.(e);
			clearFlattenedSubfolder();
		}
	} : baseProps;
	const statusLabel = itemGitStatus === "added" ? "A" : itemGitStatus === "deleted" ? "D" : itemGitStatus === "modified" ? "M" : null;
	const showStatusDot = statusLabel == null && itemContainsGitChange;
	const getItemDomId = (id) => `${treeDomId}-${id}`;
	return /* @__PURE__ */ createElement("button", {
		"data-type": "item",
		"data-item-type": hasChildren ? "folder" : "file",
		...selectionProps,
		...searchMatchProps,
		...focusedProps,
		...dragProps,
		...gitStatusProps,
		"data-item-id": itemId,
		id: getItemDomId(itemId),
		...itemProps,
		key: itemId
	}, level > 0 ? /* @__PURE__ */ jsx("div", {
		"data-item-section": "spacing",
		children: Array.from({ length: level }).map((_, index) => /* @__PURE__ */ jsx("div", {
			"data-item-section": "spacing-item",
			"data-ancestor-id": ancestors[index]
		}, index))
	}) : null, /* @__PURE__ */ jsx("div", {
		"data-item-section": "icon",
		children: hasChildren ? /* @__PURE__ */ jsx(Icon, {
			...remapIcon("file-tree-icon-chevron"),
			alignCapitals
		}) : /* @__PURE__ */ jsx(Icon, { ...remapIcon("file-tree-icon-file") })
	}), /* @__PURE__ */ jsx("div", {
		"data-item-section": "content",
		children: isFlattenedDirectory ? /* @__PURE__ */ jsx(FlattenedDirectoryName, {
			tree,
			idToPath,
			flattens: flattens ?? [],
			fallbackName: itemName
		}) : itemName
	}), statusLabel || showStatusDot ? /* @__PURE__ */ jsx("div", {
		"data-item-section": "status",
		children: statusLabel ?? /* @__PURE__ */ jsx(Icon, {
			...remapIcon("file-tree-icon-dot"),
			width: 6,
			height: 6
		})
	}) : null, isLocked ? /* @__PURE__ */ jsx("div", {
		"data-item-section": "lock",
		children: /* @__PURE__ */ jsx(Icon, { ...remapIcon("file-tree-icon-lock") })
	}) : null);
}
const TreeItem = memo(TreeItemInner, treeItemPropsAreEqual);
function Root({ fileTreeOptions, stateConfig, handleRef, callbacksRef }) {
	"use no memo";
	const { initialFiles: files, flattenEmptyDirectories, fileTreeSearchMode, gitStatus, lockedPaths, onCollision, search, sort: sortOption, useLazyDataLoader, virtualize } = fileTreeOptions;
	const iconRemap = fileTreeOptions.icons?.remap;
	const remapIcon = useCallback((name) => {
		const entry = iconRemap?.[name];
		if (entry == null) return { name };
		if (typeof entry === "string") return {
			name: entry,
			remappedFrom: name
		};
		return {
			...entry,
			remappedFrom: name
		};
	}, [iconRemap]);
	const treeDomId = useMemo(() => {
		return `ft-${(fileTreeOptions.id ?? "ft").replace(/[^A-Za-z0-9_-]/g, "_")}`;
	}, [fileTreeOptions.id]);
	const getItemDomId = (itemId) => `${treeDomId}-${itemId}`;
	const sortComparator = useMemo(() => sortOption === false ? false : sortOption != null && typeof sortOption === "object" ? sortOption.comparator : void 0, [sortOption]);
	const treeData = useMemo(() => fileListToTree(files, { sortComparator }), [files, sortComparator]);
	const { pathToId, idToPath } = useMemo(() => {
		const p2i = /* @__PURE__ */ new Map();
		const i2p = /* @__PURE__ */ new Map();
		for (const [id, node] of Object.entries(treeData)) {
			p2i.set(node.path, id);
			i2p.set(id, node.path);
		}
		return {
			pathToId: p2i,
			idToPath: i2p
		};
	}, [treeData]);
	const childToParent = useMemo(() => buildChildToParent(treeData, flattenEmptyDirectories === true), [treeData, flattenEmptyDirectories]);
	const ancestorChains = useMemo(() => buildAncestorChains(treeData, childToParent), [treeData, childToParent]);
	const restTreeConfig = useMemo(() => {
		const mapId = (item) => {
			if (treeData[item] != null) return item;
			return pathToId.get(item) ?? item;
		};
		const mapIds = (items) => {
			if (items == null) return;
			let changed = false;
			const mapped = items.map((item) => {
				const mappedItem = mapId(item);
				if (mappedItem !== item) changed = true;
				return mappedItem;
			});
			return changed ? mapped : items;
		};
		const mapState = (state$1) => {
			if (state$1 == null) return {
				state: state$1,
				changed: false
			};
			let changed = false;
			const nextState = { ...state$1 };
			const mappedExpanded = mapIds(state$1.expandedItems);
			if (mappedExpanded !== state$1.expandedItems) {
				nextState.expandedItems = mappedExpanded;
				changed = true;
			}
			const mappedSelected = mapIds(state$1.selectedItems);
			if (mappedSelected !== state$1.selectedItems) {
				nextState.selectedItems = mappedSelected;
				changed = true;
			}
			const mappedFocused = state$1.focusedItem != null ? mapId(state$1.focusedItem) : state$1.focusedItem;
			if (mappedFocused !== state$1.focusedItem) {
				nextState.focusedItem = mappedFocused;
				changed = true;
			}
			const mappedRenaming = state$1.renamingItem != null ? mapId(state$1.renamingItem) : state$1.renamingItem;
			if (mappedRenaming !== state$1.renamingItem) {
				nextState.renamingItem = mappedRenaming;
				changed = true;
			}
			const mappedChecked = mapIds(state$1.checkedItems);
			if (mappedChecked !== state$1.checkedItems) {
				nextState.checkedItems = mappedChecked;
				changed = true;
			}
			const mappedLoadingChecked = mapIds(state$1.loadingCheckPropagationItems);
			if (mappedLoadingChecked !== state$1.loadingCheckPropagationItems) {
				nextState.loadingCheckPropagationItems = mappedLoadingChecked;
				changed = true;
			}
			return {
				state: changed ? nextState : state$1,
				changed
			};
		};
		const baseConfig = {};
		const mapPathToId = (path) => {
			if (path.startsWith(FLATTENED_PREFIX)) return pathToId.get(path);
			if (flattenEmptyDirectories === true) return pathToId.get(FLATTENED_PREFIX + path) ?? pathToId.get(path);
			return pathToId.get(path);
		};
		const mapPathsToIds = (paths) => {
			if (paths == null) return void 0;
			const ids = paths.map(mapPathToId).filter((id) => id != null);
			return ids.length > 0 ? ids : [];
		};
		const topLevelInitialExpanded = stateConfig?.initialExpandedItems;
		const topLevelInitialSelected = stateConfig?.initialSelectedItems;
		const topLevelInitialSearch = stateConfig?.initialSearchQuery;
		const topLevelInitialExpandedIds = topLevelInitialExpanded != null ? expandPathsWithAncestors(topLevelInitialExpanded, pathToId, { flattenEmptyDirectories }) : void 0;
		const topLevelInitialSelectedIds = mapPathsToIds(topLevelInitialSelected);
		const mergedInitialState = topLevelInitialExpanded != null || topLevelInitialSelected != null || topLevelInitialSearch != null ? {
			...baseConfig.initialState,
			...topLevelInitialExpandedIds != null && { expandedItems: topLevelInitialExpandedIds },
			...topLevelInitialSelectedIds != null && { selectedItems: topLevelInitialSelectedIds },
			...topLevelInitialSearch != null && { search: topLevelInitialSearch }
		} : baseConfig.initialState;
		const topLevelExpanded = stateConfig?.expandedItems;
		const topLevelSelected = stateConfig?.selectedItems;
		const topLevelExpandedIds = topLevelExpanded != null ? controlledExpandedPathsToExpandedIds(topLevelExpanded, pathToId, { flattenEmptyDirectories }) : void 0;
		const topLevelSelectedIds = mapPathsToIds(topLevelSelected);
		const mergedState = topLevelExpanded != null || topLevelSelected != null ? {
			...baseConfig.state,
			...topLevelExpandedIds != null && { expandedItems: topLevelExpandedIds },
			...topLevelSelectedIds != null && { selectedItems: topLevelSelectedIds }
		} : baseConfig.state;
		const configWithMergedState = {
			...baseConfig,
			...mergedInitialState != null && { initialState: mergedInitialState },
			...mergedState != null && { state: mergedState }
		};
		const initialState = mapState(configWithMergedState.initialState);
		const state = mapState(configWithMergedState.state);
		if (!initialState.changed && !state.changed) return configWithMergedState;
		return {
			...configWithMergedState,
			...initialState.state != null && { initialState: initialState.state },
			...state.state != null && { state: state.state }
		};
	}, [
		treeData,
		pathToId,
		stateConfig,
		flattenEmptyDirectories
	]);
	const dataLoader = useMemo(() => useLazyDataLoader === true ? generateLazyDataLoader(files, {
		flattenEmptyDirectories,
		sortComparator
	}) : generateSyncDataLoader(files, {
		flattenEmptyDirectories,
		sortComparator
	}), [
		files,
		flattenEmptyDirectories,
		useLazyDataLoader,
		sortComparator
	]);
	const isDnD = fileTreeOptions.dragAndDrop === true;
	const features = useMemo(() => {
		const base = [
			syncDataLoaderFeature,
			selectionFeature,
			hotkeysCoreFeature,
			fileTreeSearchFeature,
			expandAllFeature,
			gitStatusFeature
		];
		if (isDnD) base.push(dragAndDropFeature, keyboardDragAndDropFeature);
		base.push(propMemoizationFeature);
		return base;
	}, [isDnD]);
	const filesRef = useRef(files);
	filesRef.current = files;
	const flattenedDropSubfolderIdRef = useRef(null);
	const flattenedHighlightRef = useRef(null);
	const detectFlattenedSubfolder = useCallback((e) => {
		let el = e.target;
		if (el != null && el.nodeType === Node.TEXT_NODE) el = el.parentElement;
		const span = el?.closest?.("[data-item-flattened-subitem]");
		const id = span?.getAttribute("data-item-flattened-subitem") ?? null;
		if (id === flattenedDropSubfolderIdRef.current) return;
		if (flattenedHighlightRef.current != null) flattenedHighlightRef.current.removeAttribute("data-item-flattened-subitem-drag-target");
		if (span != null && id != null) {
			span.setAttribute("data-item-flattened-subitem-drag-target", "true");
			flattenedHighlightRef.current = span;
			flattenedDropSubfolderIdRef.current = id;
		} else {
			flattenedHighlightRef.current = null;
			flattenedDropSubfolderIdRef.current = null;
		}
	}, []);
	const clearFlattenedSubfolder = useCallback(() => {
		if (flattenedHighlightRef.current != null) flattenedHighlightRef.current.removeAttribute("data-item-flattened-subitem-drag-target");
		flattenedHighlightRef.current = null;
		flattenedDropSubfolderIdRef.current = null;
	}, []);
	const detectFlattenedSubfolderFromPoint = useCallback((clientX, clientY) => {
		let el = (tree.getElement()?.getRootNode() ?? document).elementFromPoint(clientX, clientY);
		if (el != null && el.nodeType === Node.TEXT_NODE) el = el.parentElement;
		const span = el?.closest?.("[data-item-flattened-subitem]");
		const id = span?.getAttribute("data-item-flattened-subitem") ?? null;
		if (id === flattenedDropSubfolderIdRef.current) return;
		if (flattenedHighlightRef.current != null) flattenedHighlightRef.current.removeAttribute("data-item-flattened-subitem-drag-target");
		if (span != null && id != null) {
			span.setAttribute("data-item-flattened-subitem-drag-target", "true");
			flattenedHighlightRef.current = span;
			flattenedDropSubfolderIdRef.current = id;
		} else {
			flattenedHighlightRef.current = null;
			flattenedDropSubfolderIdRef.current = null;
		}
	}, []);
	const prevIdToPathRef = useRef(idToPath);
	const pendingDropTargetExpandRef = useRef(null);
	const onDropHandler = useCallback((items, target) => {
		const draggedPaths = items.map((item) => item.getItemData().path);
		let targetPath = target.item.getId() === "root" ? "root" : target.item.getItemData().path;
		if (flattenedDropSubfolderIdRef.current != null) {
			targetPath = idToPath.get(flattenedDropSubfolderIdRef.current) ?? targetPath;
			flattenedDropSubfolderIdRef.current = null;
		}
		const newFiles = computeNewFilesAfterDrop(filesRef.current, draggedPaths, targetPath, { onCollision });
		if (targetPath !== "root") pendingDropTargetExpandRef.current = {
			path: targetPath.startsWith(FLATTENED_PREFIX) ? targetPath.slice(FLATTENED_PREFIX.length) : targetPath,
			expectedFilesSignature: getFilesSignature(newFiles)
		};
		else pendingDropTargetExpandRef.current = null;
		callbacksRef?.current._onDragMoveFiles?.(newFiles);
	}, [
		callbacksRef,
		onCollision,
		idToPath
	]);
	const searchActiveRef = useRef(false);
	const searchModeConfig = {
		fileTreeSearchMode,
		search
	};
	const gitStatusConfig = {
		gitStatus,
		gitStatusSignature: getGitStatusSignature(gitStatus),
		gitStatusPathToId: pathToId
	};
	const tree = useTree({
		...restTreeConfig,
		...searchModeConfig,
		...gitStatusConfig,
		rootItemId: "root",
		dataLoader,
		getItemName: (item) => item.getItemData().name,
		isItemFolder: (item) => {
			return item.getItemData()?.children?.direct != null;
		},
		hotkeys: {
			customExpandAll: {
				hotkey: "KeyQ",
				handler: (_e, tree$1) => {
					tree$1.expandAll();
				}
			},
			customCollapseAll: {
				hotkey: "KeyW",
				handler: (_e, tree$1) => {
					tree$1.collapseAll();
				}
			}
		},
		features,
		...isDnD && {
			canReorder: false,
			canDrag: (items) => {
				if (searchActiveRef.current) return false;
				if (lockedPaths == null || lockedPaths.length === 0) return true;
				const lockedSet = new Set(lockedPaths);
				for (const item of items) {
					const path = item.getItemData().path;
					if (path != null && lockedSet.has(getSelectionPath(path))) return false;
				}
				return true;
			},
			onDrop: onDropHandler,
			canDrop: (_items, target) => target.item.isFolder(),
			openOnDropDelay: 800,
			_onTouchDragMove: detectFlattenedSubfolderFromPoint,
			_onTouchDragEnd: clearFlattenedSubfolder
		}
	});
	searchActiveRef.current = (tree.getState().search?.length ?? 0) > 0;
	const pendingExpandMigrationRef = useRef(null);
	if (prevIdToPathRef.current !== idToPath) {
		const currentExpandedIds = tree.getState().expandedItems ?? [];
		if (currentExpandedIds.some((id) => !idToPath.has(id))) {
			const oldIdToPath = prevIdToPathRef.current;
			pendingExpandMigrationRef.current = currentExpandedIds.map((id) => oldIdToPath.get(id)).filter((p) => p != null).map((p) => p.startsWith(FLATTENED_PREFIX) ? p.slice(FLATTENED_PREFIX.length) : p);
		}
	}
	prevIdToPathRef.current = idToPath;
	useEffect(() => {
		if (handleRef == null) return;
		handleRef.current = {
			tree,
			pathToId,
			idToPath
		};
		return () => {
			handleRef.current = null;
		};
	}, [
		tree,
		pathToId,
		idToPath,
		handleRef
	]);
	useEffect(() => {
		const previousPaths = pendingExpandMigrationRef.current;
		const pendingDropTarget = pendingDropTargetExpandRef.current;
		const dropTarget = pendingDropTarget != null && pendingDropTarget.expectedFilesSignature === getFilesSignature(files) ? pendingDropTarget.path : null;
		pendingExpandMigrationRef.current = null;
		pendingDropTargetExpandRef.current = null;
		if (previousPaths == null && dropTarget == null) return;
		const pathsToExpand = previousPaths != null ? [...previousPaths] : [];
		if (dropTarget != null) pathsToExpand.push(dropTarget);
		const expandIds = expandPathsWithAncestors(pathsToExpand, pathToId, { flattenEmptyDirectories });
		if (previousPaths != null) tree.applySubStateUpdate("expandedItems", () => expandIds);
		else {
			const currentExpanded = tree.getState().expandedItems ?? [];
			const currentSet = new Set(currentExpanded);
			const newIds = expandIds.filter((id) => !currentSet.has(id));
			if (newIds.length === 0) return;
			tree.applySubStateUpdate("expandedItems", (prev) => [...prev ?? [], ...newIds]);
		}
		tree.rebuildTree();
	}, [
		files,
		pathToId,
		tree,
		flattenEmptyDirectories
	]);
	const selectionSnapshotRef = useRef(null);
	const selectionSnapshot = tree.getState().selectedItems?.join("|") ?? "";
	useEffect(() => {
		const onSelection = callbacksRef?.current.onSelection;
		if (onSelection == null) return;
		if (selectionSnapshotRef.current == null) {
			selectionSnapshotRef.current = selectionSnapshot;
			return;
		}
		if (selectionSnapshotRef.current === selectionSnapshot) return;
		selectionSnapshotRef.current = selectionSnapshot;
		onSelection(tree.getSelectedItems().map((item) => {
			const data = item.getItemData();
			return {
				path: getSelectionPath(data.path),
				isFolder: data.children?.direct != null
			};
		}));
	}, [
		selectionSnapshot,
		callbacksRef,
		tree
	]);
	const expandedSnapshotRef = useRef(null);
	const expandedSnapshot = tree.getState().expandedItems?.join("|") ?? "";
	useEffect(() => {
		const onExpandedItemsChange = callbacksRef?.current.onExpandedItemsChange;
		if (onExpandedItemsChange == null) return;
		if (expandedSnapshotRef.current == null) {
			expandedSnapshotRef.current = expandedSnapshot;
			return;
		}
		if (expandedSnapshotRef.current === expandedSnapshot) return;
		expandedSnapshotRef.current = expandedSnapshot;
		const ids = tree.getState().expandedItems ?? [];
		onExpandedItemsChange(filterOrphanedPaths([...new Set(ids.map((id) => idToPath.get(id)).filter((path) => path != null).map(getSelectionPath))], pathToId, flattenEmptyDirectories));
	}, [
		expandedSnapshot,
		callbacksRef,
		tree,
		idToPath,
		pathToId,
		flattenEmptyDirectories
	]);
	const selectedSnapshotRef = useRef(null);
	const selectedSnapshot = tree.getState().selectedItems?.join("|") ?? "";
	useEffect(() => {
		const onSelectedItemsChange = callbacksRef?.current.onSelectedItemsChange;
		if (onSelectedItemsChange == null) return;
		if (selectedSnapshotRef.current == null) {
			selectedSnapshotRef.current = selectedSnapshot;
			return;
		}
		if (selectedSnapshotRef.current === selectedSnapshot) return;
		selectedSnapshotRef.current = selectedSnapshot;
		onSelectedItemsChange((tree.getState().selectedItems ?? []).map((id) => idToPath.get(id)).filter((path) => path != null).map(getSelectionPath));
	}, [
		selectedSnapshot,
		callbacksRef,
		tree,
		idToPath
	]);
	useLayoutEffect(() => {
		const search$1 = tree.getState().search;
		if (search$1 != null && search$1.length > 0) tree.setSearch(search$1);
	}, [tree]);
	const { onChange,...origSearchInputProps } = tree.getSearchInputElementProps();
	const shouldRenderSearchInput = search === true;
	const hasFocusedItem = tree.getState().focusedItem != null;
	const focusedItemId = hasFocusedItem ? tree.getState().focusedItem : null;
	const activeDescendantId = (tree.isSearchOpen?.() ?? false) && focusedItemId != null ? getItemDomId(focusedItemId) : void 0;
	const searchInputProps = {
		...origSearchInputProps,
		...activeDescendantId != null && {
			"aria-activedescendant": activeDescendantId,
			"aria-controls": treeDomId
		},
		onInput: onChange
	};
	const guideStyleText = useMemo(() => {
		const selectedIds = tree.getState().selectedItems ?? [];
		if (selectedIds.length === 0 && focusedItemId == null) return "";
		const parentIds = /* @__PURE__ */ new Set();
		for (const id of selectedIds) {
			const parentId = childToParent.get(id);
			if (parentId != null && parentId !== "root") parentIds.add(parentId);
		}
		if (focusedItemId != null) {
			const focusedParentId = childToParent.get(focusedItemId);
			if (focusedParentId != null && focusedParentId !== "root") parentIds.add(focusedParentId);
		}
		if (parentIds.size === 0) return "";
		const escape = (v) => v.replace(/\\/g, "\\\\").replace(/"/g, "\\\"");
		return `:is(${Array.from(parentIds).map((id) => `[data-item-section="spacing-item"][data-ancestor-id="${escape(id)}"]`).join(",\n")}) { opacity: 1; }`;
	}, [
		selectionSnapshot,
		focusedItemId,
		childToParent
	]);
	const shouldVirtualize = virtualize != null && virtualize !== false;
	const virtualizeThreshold = shouldVirtualize ? Math.max(0, virtualize.threshold) : Number.POSITIVE_INFINITY;
	return /* @__PURE__ */ jsxs("div", {
		...tree.getContainerProps(),
		id: treeDomId,
		"data-file-tree-virtualized-root": shouldVirtualize ? "true" : void 0,
		children: [
			/* @__PURE__ */ jsx("style", { dangerouslySetInnerHTML: { __html: guideStyleText } }),
			shouldRenderSearchInput ? /* @__PURE__ */ jsx("div", {
				"data-file-tree-search-container": true,
				children: /* @__PURE__ */ jsx("input", {
					placeholder: "Search…",
					"data-file-tree-search-input": true,
					...searchInputProps
				})
			}) : null,
			(() => {
				const allItems = tree.getItems();
				const visibleIdSet = getSearchVisibleIdSet(tree);
				const gitStatusMap = getGitStatusMap(tree);
				const items = visibleIdSet != null ? allItems.filter((item) => visibleIdSet.has(item.getId())) : allItems;
				const lockedPathSet = lockedPaths != null && lockedPaths.length > 0 ? new Set(lockedPaths) : null;
				const renderItemAtIndex = (index) => {
					const item = items[index];
					if (item == null) return null;
					const itemData = item.getItemData();
					const itemMeta = item.getItemMeta();
					const hasChildren = itemData?.children?.direct != null;
					const isExpanded = hasChildren && item.isExpanded();
					const itemName = item.getItemName();
					const level = itemMeta.level;
					const itemPath = itemData?.path;
					const isLocked = itemPath != null && lockedPathSet?.has(getSelectionPath(itemPath)) === true;
					const isSelected = item.isSelected();
					const isFlattenedDirectory = itemData?.flattens != null;
					const isSearchMatch = item.isMatchingSearch();
					const isFocused = hasFocusedItem && item.isFocused();
					const isDragTarget = isDnD && item.isUnorderedDragTarget?.() === true;
					const isDragging = isDnD && tree.getState().dnd?.draggedItems?.some((d) => d.getId() === item.getId()) === true;
					const itemGitStatus = gitStatusMap?.statusById.get(item.getId());
					const itemContainsGitChange = hasChildren && (gitStatusMap?.foldersWithChanges.has(item.getId()) ?? false);
					const ancestors = ancestorChains.get(item.getId()) ?? EMPTY_ANCESTORS;
					return /* @__PURE__ */ jsx(TreeItem, {
						item,
						tree,
						itemId: item.getId(),
						hasChildren,
						isExpanded,
						itemName,
						level,
						isSelected,
						isFocused,
						isSearchMatch,
						isDragTarget,
						isDragging,
						isDnD,
						isFlattenedDirectory,
						isLocked,
						gitStatus: itemGitStatus,
						containsGitChange: itemContainsGitChange ?? false,
						flattens: itemData?.flattens,
						idToPath,
						ancestors,
						treeDomId,
						remapIcon,
						detectFlattenedSubfolder,
						clearFlattenedSubfolder
					}, item.getId());
				};
				if (shouldVirtualize && items.length > 0 && items.length >= virtualizeThreshold) {
					const focusedIndex = focusedItemId != null ? items.findIndex((item) => item.getId() === focusedItemId) : null;
					return /* @__PURE__ */ jsx("div", {
						"data-file-tree-virtualized-scroll": "true",
						children: /* @__PURE__ */ jsx(VirtualizedList, {
							itemCount: items.length,
							renderItem: renderItemAtIndex,
							scrollToIndex: focusedIndex != null && focusedIndex >= 0 ? focusedIndex : null
						})
					});
				}
				return items.map((_, i) => renderItemAtIndex(i));
			})()
		]
	});
}

//#endregion
export { Root };
