// @pierre/trees@0.0.1-beta.4/dist/components/VirtualizedList.js
import { useLayoutEffect, useRef, useState } from "preact/hooks";
import { jsx, jsxs } from "preact/jsx-runtime";

//#region src/components/VirtualizedList.tsx
const OVERSCAN = 10;
const DEFAULT_ITEM_HEIGHT = 30;
const EMPTY_RANGE = {
	start: 0,
	end: -1
};
/**
* Walk up from `el` (crossing shadow boundaries) to find the nearest ancestor
* with `overflow-y: auto | scroll`. Falls back to `document.documentElement`.
*/
function findScrollableAncestor(el) {
	let node = el.parentElement;
	while (node != null) {
		const style = getComputedStyle(node);
		if (style.overflowY === "auto" || style.overflowY === "scroll" || style.overflow === "auto" || style.overflow === "scroll") return node;
		if (node.parentElement != null) node = node.parentElement;
		else {
			const root = node.getRootNode();
			if (root instanceof ShadowRoot) node = root.host;
			else break;
		}
	}
	return document.documentElement;
}
function resolveItemHeight(container, explicitItemHeight) {
	if (explicitItemHeight != null && explicitItemHeight > 0) return explicitItemHeight;
	const cssValue = getComputedStyle(container).getPropertyValue("--ft-internal-row-height").trim();
	const parsed = Number.parseFloat(cssValue);
	return Number.isFinite(parsed) && parsed > 0 ? parsed : DEFAULT_ITEM_HEIGHT;
}
function getOffsetWithinViewport(container, viewport) {
	let top = 0;
	let node = container;
	while (node != null && node !== viewport) {
		top += node.offsetTop;
		node = node.offsetParent;
	}
	if (node === viewport) return top;
	return container.getBoundingClientRect().top - viewport.getBoundingClientRect().top + viewport.scrollTop;
}
function normalizeRange(range, itemCount) {
	if (itemCount <= 0 || range.end < range.start) return EMPTY_RANGE;
	const start = Math.max(0, Math.min(range.start, itemCount - 1));
	return {
		start,
		end: Math.max(start, Math.min(range.end, itemCount - 1))
	};
}
function rangesEqual(a, b) {
	return a.start === b.start && a.end === b.end;
}
function computeVisibleRange({ scrollTop, viewportHeight, offset, itemCount, itemHeight }) {
	if (itemCount <= 0) return EMPTY_RANGE;
	const rawStart = Math.floor((scrollTop - offset) / itemHeight);
	const rawEnd = Math.ceil((scrollTop - offset + viewportHeight) / itemHeight) - 1;
	if (rawEnd < 0 || rawStart >= itemCount) return EMPTY_RANGE;
	return {
		start: Math.max(0, rawStart),
		end: Math.min(itemCount - 1, rawEnd)
	};
}
function expandRange(range, itemCount, overscan) {
	if (range.end < range.start || itemCount <= 0) return EMPTY_RANGE;
	return normalizeRange({
		start: range.start - overscan,
		end: range.end + overscan
	}, itemCount);
}
function computeWindowRange(metrics, currentRange = EMPTY_RANGE) {
	const visibleRange = computeVisibleRange(metrics);
	const normalizedCurrentRange = normalizeRange(currentRange, metrics.itemCount);
	if (normalizedCurrentRange.end >= normalizedCurrentRange.start && visibleRange.start >= normalizedCurrentRange.start && visibleRange.end <= normalizedCurrentRange.end) return normalizedCurrentRange;
	return expandRange(visibleRange, metrics.itemCount, metrics.overscan ?? OVERSCAN);
}
function computeStickyWindowLayout({ range, itemCount, itemHeight, viewportHeight }) {
	const totalHeight = Math.max(0, itemCount * itemHeight);
	if (range.end < range.start) return {
		totalHeight,
		offsetHeight: 0,
		windowHeight: 0,
		stickyInset: 0
	};
	const offsetHeight = range.start * itemHeight;
	const windowHeight = (range.end - range.start + 1) * itemHeight;
	return {
		totalHeight,
		offsetHeight,
		windowHeight,
		stickyInset: Math.min(0, viewportHeight - windowHeight)
	};
}
function VirtualizedList({ itemCount, renderItem, scrollToIndex, itemHeight }) {
	"use no memo";
	const containerRef = useRef(null);
	const stickyOffsetRef = useRef(null);
	const stickyWindowRef = useRef(null);
	const viewportRef = useRef(null);
	const [range, setRange] = useState(EMPTY_RANGE);
	const [resolvedHeight, setResolvedHeight] = useState(itemHeight != null && itemHeight > 0 ? itemHeight : DEFAULT_ITEM_HEIGHT);
	const [viewportHeight, setViewportHeight] = useState(0);
	useLayoutEffect(() => {
		const container = containerRef.current;
		if (container == null) return;
		const viewport = container.parentElement ?? findScrollableAncestor(container);
		viewportRef.current = viewport;
		let scrollTimer = null;
		const update = () => {
			const nextHeight = resolveItemHeight(container, itemHeight);
			const nextViewportHeight = viewport.clientHeight;
			const offset = getOffsetWithinViewport(container, viewport);
			setResolvedHeight((prev) => prev === nextHeight ? prev : nextHeight);
			setViewportHeight((prev) => prev === nextViewportHeight ? prev : nextViewportHeight);
			setRange((prev) => {
				const next = computeWindowRange({
					scrollTop: viewport.scrollTop,
					viewportHeight: nextViewportHeight,
					offset,
					itemCount,
					itemHeight: nextHeight
				}, prev);
				return rangesEqual(prev, next) ? prev : next;
			});
		};
		const onScroll = () => {
			update();
			container.dataset.isScrolling ??= "";
			if (scrollTimer != null) clearTimeout(scrollTimer);
			scrollTimer = setTimeout(() => {
				delete container.dataset.isScrolling;
				scrollTimer = null;
			}, 50);
		};
		update();
		viewport.addEventListener("scroll", onScroll, { passive: true });
		const ro = typeof ResizeObserver !== "undefined" ? new ResizeObserver(update) : null;
		ro?.observe(viewport);
		return () => {
			viewport.removeEventListener("scroll", onScroll);
			if (scrollTimer != null) clearTimeout(scrollTimer);
			delete container.dataset.isScrolling;
			ro?.disconnect();
		};
	}, [itemCount, itemHeight]);
	useLayoutEffect(() => {
		if (scrollToIndex == null || scrollToIndex < 0) return;
		const viewport = viewportRef.current;
		const container = containerRef.current;
		if (viewport == null || container == null) return;
		const offset = getOffsetWithinViewport(container, viewport);
		const itemTop = offset + scrollToIndex * resolvedHeight;
		const itemBottom = itemTop + resolvedHeight;
		const viewTop = viewport.scrollTop;
		const viewBottom = viewTop + viewport.clientHeight;
		let nextScrollTop = viewTop;
		if (itemTop < viewTop) nextScrollTop = itemTop;
		else if (itemBottom > viewBottom) nextScrollTop = itemBottom - viewport.clientHeight;
		if (nextScrollTop !== viewTop) {
			setRange((prev) => {
				const next = computeWindowRange({
					scrollTop: nextScrollTop,
					viewportHeight: viewport.clientHeight,
					offset,
					itemCount,
					itemHeight: resolvedHeight
				});
				return rangesEqual(prev, next) ? prev : next;
			});
			viewport.scrollTop = nextScrollTop;
		}
	}, [
		scrollToIndex,
		resolvedHeight,
		itemCount
	]);
	const { totalHeight, offsetHeight, windowHeight, stickyInset } = computeStickyWindowLayout({
		range,
		itemCount,
		itemHeight: resolvedHeight,
		viewportHeight
	});
	useLayoutEffect(() => {
		const container = containerRef.current;
		if (container != null) container.style.height = `${totalHeight}px`;
		const stickyOffset = stickyOffsetRef.current;
		if (stickyOffset != null) stickyOffset.style.height = `${offsetHeight}px`;
		const stickyWindow = stickyWindowRef.current;
		if (stickyWindow != null) {
			stickyWindow.style.height = `${windowHeight}px`;
			stickyWindow.style.top = `${stickyInset}px`;
			stickyWindow.style.bottom = `${stickyInset}px`;
		}
	}, [
		totalHeight,
		offsetHeight,
		windowHeight,
		stickyInset
	]);
	const { start: startIndex, end: endIndex } = range;
	const children = [];
	for (let i = startIndex; i <= endIndex; i++) {
		const item = renderItem(i);
		if (item != null) children.push(item);
	}
	return /* @__PURE__ */ jsxs("div", {
		ref: containerRef,
		"data-file-tree-virtualized-list": "true",
		children: [/* @__PURE__ */ jsx("div", {
			ref: stickyOffsetRef,
			"data-file-tree-virtualized-sticky-offset": "true",
			"aria-hidden": "true"
		}), /* @__PURE__ */ jsx("div", {
			ref: stickyWindowRef,
			"data-file-tree-virtualized-sticky": "true",
			children
		})]
	});
}

//#endregion
export { VirtualizedList, computeStickyWindowLayout, computeVisibleRange, computeWindowRange };
